<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>TODO • multienrichjam</title><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="TODO"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">multienrichjam</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.105.900</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="articles/importClusterProfiler.html">Multienrichment with clusterProfiler</a></li>
    <li><a class="dropdown-item" href="articles/importIPA.html">How to use Ingenuity IPA enrichment results</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Explanation</h6></li>
    <li><a class="dropdown-item" href="articles/MemPlotConcepts.html">Mem Plot Concepts</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="http://github.com/jmw86069/multienrichjam/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>TODO</h1>
      <small class="dont-index">Source: <a href="http://github.com/jmw86069/multienrichjam/blob/HEAD/TODO.md" class="external-link"><code>TODO.md</code></a></small>
    </div>

<div id="todo" class="section level1">

<div class="section level2">
<h2 id="id_24nov2025">24nov2025<a class="anchor" aria-label="anchor" href="#id_24nov2025"></a></h2>
<ul><li>
<p><code>MemPlotFolio</code> enhancements</p>
<ul><li>Make it so <code>GenePathHeatmap()</code> and <code>EnrichmentHeatmap()</code> can dynamically re-create the heatmap, so customizations can be made without re-running <code><a href="reference/mem_plot_folio.html">prepare_folio()</a></code>?</li>
</ul></li>
<li>
<p><code>plot_cnet_heatmaps()</code>, <code>CnetHeatmaps()</code></p>
<ul><li>Consider “dev” mode that prints some unit grid onto the plot, to help align labels and adjust margins.</li>
<li>requires expression data, column groups</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_18nov2025">18nov2025<a class="anchor" aria-label="anchor" href="#id_18nov2025"></a></h2>
<ul><li>
<p>Usability / bugs</p>
<ul><li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>FIXED. Bug: <code>list</code> for <code>pathway_column_split</code> with subset pathways; it seems to have clusters as <code>character</code> with this approach.</li>
<li>FIXED. applies <code>simple_anno_size</code> to row and not column in gene-path heatmap.</li>
<li>DONE. Make ‘mpf’ argument also use <code>Clusters()</code> and <code>GeneClusters()</code>. May need to review whether this behavior is optimal for genes especially.</li>
</ul></li>
<li><p>FIXED. <code><a href="reference/fixSetLabels.html">fixSetLabels()</a></code> add option for max width, truncating anything longer instead of applying word-wrap. (Already available, haha <code>maxNchar</code>.)</p></li>
<li><p>DONE. <code><a href="reference/mem2cnet.html">mem2cnet()</a></code> consider aesthetic defaults: <code>vertex.label.font=2</code>, <code>use_shadowText=TRUE</code> (store in attributes?)</p></li>
<li><p>DONE. Change default <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code> simple_anno_size to 5mm.</p></li>
<li><p>UNSURE. Consider new default cluster_column_slices=TRUE, and for rows. It extends the dendrograms with connections across slices.</p></li>
</ul></li>
<li>
<p>Consider new function <code>merge_Mem_sets()</code></p>
<ul><li>intended to combine “equivalent” pathways (TLR signaling v1, TLR signaling v2) together.</li>
<li>Bonus points: Pathways with identical names (v1, v2, etc) with greater than threshold (75%?) identical genes, automatically merge them.</li>
</ul></li>
<li><p><code>CnetCollapsed()</code> need some way to adjust layout, and apply reorder, spread labels, etc.</p></li>
<li><p><code>CnetExemplar()</code> consider adding pathway title as prefix, for example “A: Pleural Mesothelioma”, to indicate the cluster for each exemplar.</p></li>
<li><p><code><a href="reference/removeIgraphSinglets.html">removeIgraphSinglets()</a></code> should subset graph_attr(“mark.groups”).</p></li>
<li>
<p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code></p>
<ul><li>with <code>mark.groups=TRUE</code> and non-Cnet, it should try <code>components()</code>, however using <code>mark.groups=igraph::components</code> works currently.</li>
</ul></li>
<li>
<p><code>make_alpha_hull()</code></p>
<ul><li>try to split extremely disparate node groups, e.g. two sides of the plot, should ideally split into two subcomponents.</li>
<li>draw hull labels slightly farther from the edges? It might need something helpful from <code><a href="https://jmw86069.github.io/jamba/reference/drawLabels.html" class="external-link">jamba::drawLabels()</a></code>.</li>
</ul></li>
<li>
<p>Usability / Workflows</p>
<ul><li>Consider workflow to re-Mem a subset of pathway clusters. Situation: Pathway clusters often have one “miscellaneous” cluster with few genes, no coherence. Target: Re-create Mem without this cluster. <code>prepare_folio(Mem, pathway_column_split=Clusters(Mpf)[c(1, 3, 4)])</code>
</li>
</ul></li>
<li><p>Ideal world: <code>GenePathHeatmap(Mpf[, clusters])</code> would use the subset.</p></li>
<li><p>Debug <code><a href="reference/rank_mem_clusters.html">rank_mem_clusters()</a></code> for follow-up with using Mem/Mpf input.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_14nov2025">14nov2025<a class="anchor" aria-label="anchor" href="#id_14nov2025"></a></h2>
<ul><li><p>Add vignette for <code><a href="reference/launch_shinycat.html">launch_shinycat()</a></code></p></li>
<li>
<p>Add <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind()</a></code> and <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code> to combine multiple Mem objects.</p>
<ul><li>Use case: enrichment with IPA; enrichment with clusterProfiler; then <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code> them.</li>
<li>Use case: enrichment of A,B in IPA; enrichment of C,D in IPA; then <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind()</a></code> them.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12nov2025">12nov2025<a class="anchor" aria-label="anchor" href="#id_12nov2025"></a></h2>
<ul><li>
<p>Continue polishing the workflow:</p>
<ul><li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>Accept optional <code>MemPlotFolio</code> input, to re-use identical settings.</li>
<li>consider new <code>PrepareFolio()</code> with <code>do_plot=FALSE</code>?<br>
Workflow: <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> <code>PrepareFolio()</code> individual plots: <code>EnrichmentHeatmap()</code>, <code>GenePathHeatmap()</code>, etc. <code>PlotFolio()</code>
</li>
<li>new arg <code>do_whichbyname</code> accept <code>character</code> string?</li>
<li>consider replacing plots 3,4,5 with one Cnet, then move the custom label into <code>CnetCollapsed()</code>
</li>
</ul></li>
<li><p><code>CnetHeatmap()</code>, <code>GenePathHeatmap()</code> could use <code>Mem</code> input? Secretly call <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> behind the scenes.</p></li>
</ul></li>
<li>
<p>COMPLETE. Move <code><a href="reference/fixSetLabels.html">fixSetLabels()</a></code> logic to config file which is prepared upon loading the package.</p>
<ul><li>COMPLETE. words_from, words_to; add_from, add_to; abbrev_from, abbrev_to: Should they be <code>data.frame</code> objects?</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10nov2025">10nov2025<a class="anchor" aria-label="anchor" href="#id_10nov2025"></a></h2>
<ul><li>
<p>COMPLETE. <code><a href="reference/IPAlist_to_hits.html">IPAlist_to_hits()</a></code> to extract hit list from IPA xref tables.</p>
<ul><li>convert “/” to “:” for genes with two or more symbols</li>
<li>remove <code>" (includes others)"</code> suffix for proper gene symbol matching</li>
</ul></li>
<li>
<p>COMPLETE. <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>Replace CnetCollapsed with one <code>igraph</code> that updates <code>V(cnet)$label</code> as needed. The rest should be constant.</li>
</ul></li>
<li><p>Add <code>PlotCnetHeatmaps()</code></p></li>
<li>
<p><code><a href="reference/colorRamp2D.html">colorRamp2D()</a></code> enhancements</p>
<ul><li>Use <code>thresholds(Mem)$direction_cutoff</code>, add labeling</li>
<li>consider continuous color option, also use <code>direction_cutoff</code>
</li>
</ul></li>
<li><p>Consider making many functions “internal” to simplify the user experience.</p></li>
</ul><div class="section level3">
<h3 id="new-ideas">New ideas<a class="anchor" aria-label="anchor" href="#new-ideas"></a></h3>
<ul><li>COMPLETE. Consider some way to score gene clusters by how well they “define” the cluster? For example, some genes are found in the majority of nodes in at least one cluster. Other genes are scattered across clusters with no visible aggregation. The metric could be fraction of genes in at least half the pathways in one pathway cluster.</li>
</ul></div>
</div>
<div class="section level2">
<h2 id="id_04nov2025">04nov2025<a class="anchor" aria-label="anchor" href="#id_04nov2025"></a></h2>
<ul><li>
<p>PARTIAL. Simplify the function reference categories.</p>
<ul><li>DONE. Move core user-facing functions to top categories.</li>
<li>Move almost everything into “Details”.</li>
</ul></li>
<li><p>Split vignettes into concise subtopics.</p></li>
<li><p>COMPLETE. Consider <code><a href="https://rdrr.io/pkg/Biobase/man/class.Versioned.html" class="external-link">Biobase::Versioned</a></code> to version <code>Mem</code> class structure.</p></li>
<li><p>COMPLETE. Write <code>updateObject()</code>.</p></li>
<li>
<p>Make some Mem slots optional, consider gradual reduction of slots?</p>
<ul><li>Use slot type “matrix_OR_NULL” with <code>setClassUnion("matrix_OR_NULL", c("matrix", "NULL"))</code>.</li>
<li>The <code>geneIMcolors()</code> accessor would return geneIMcolors when defined, or generate the color matrix dynamically using thresholds and colorV.</li>
<li>Candidates: geneIMcolors, enrichIMcolors, geneIMdirection multiEnrichDF, multiEnrichList</li>
</ul></li>
<li><p>Add <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind()</a></code> and <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind()</a></code> to combine multiple Mem objects.</p></li>
<li><p>Consider providing some data dynamically, using current thresholds: <code>enrichIMcolors</code>, <code>geneIMcolors</code></p></li>
<li><p>Add <code>plot_cnet_heatmaps()</code>. It needs pathway and expression data to test.</p></li>
<li>
<p>COMPLETE. Add <code>MemPlotFolio</code> S4 object</p>
<ul><li>Store data in <code>plotdata</code> as a <code>list</code> or <code>SimpleList</code>
</li>
<li>Store parameters in a slot <code>parameters</code> as <code>list</code> as well.</li>
<li>Accessors: GPHeatmap, EHeatmap, Cnet</li>
<li>In theory, all data could be generated via accessors, and just store the parameters.</li>
</ul></li>
<li>
<p>Consider serialization functions</p>
<ul><li>save series of incident matrices, data.frame objects</li>
<li>confirm it can be re-loaded directly to <code>Mem</code> object</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_29oct2025">29oct2025<a class="anchor" aria-label="anchor" href="#id_29oct2025"></a></h2>
<ul><li>
<p>Mem S4 object updates:</p>
<ul><li><p>Consider better mechanism to rename pathway gene sets with <code>sets()&lt;-</code>, specifically so that the <code>enrichList</code> results are also kept consistent. Perhaps an alias to change from original name to custom name?</p></li>
<li><p>Consider auto-updating <code>Mem</code> when parameters/thresholds change.</p></li>
<li>
<p>Consider removing some slots and replacing the accessor with function:</p>
<ul><li>
<code>geneIMcolors</code> and <code>enrichIMcolors</code> could be done using with dynamic method which applies <code>colorV</code>,<code>p_cutoff</code>,<code>min_count</code> to produce the color matrix. Would enable changing thresholds and having up-to-date data. Also would remove the slot and reduce S4 complexity.</li>
<li>
<code>multiEnrichDF</code>,<code>multiEnrichResult</code>
</li>
</ul></li>
</ul></li>
<li>
<p>Consider S4 classes:</p>
<ul><li><code>MemPlotFolio</code></li>
<li>
<code>multiEnrichResult</code> - simple extension to <code>enrichResult</code>?</li>
<li>
<code>enrichResultList</code> - need to determine if it enables something useful, other than just making it clear the type of data</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_23oct2025">23oct2025<a class="anchor" aria-label="anchor" href="#id_23oct2025"></a></h2>
<ul><li>
<p>Organize vignettes using diataxis philosophy</p>
<ul><li><p>Tutorial: Use IPA; Use clusterProfiler</p></li>
<li>
<p>How-to:</p>
<ul><li>Define pathway-gene clustering parameters</li>
<li>Customize Mem Enrichment Heatmap</li>
<li>Adjust Cnet plot layout</li>
</ul></li>
<li>
<p>Explanation:</p>
<ul><li>Determine appropriate parameters for <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>
</li>
<li>Decide Cnet strategy for a visual summary</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_17oct2025">17oct2025<a class="anchor" aria-label="anchor" href="#id_17oct2025"></a></h2>
<ul><li><p>DONE. Remove all dependencies on arules::transactions</p></li>
<li><p>Clean up the IPA vignette figure sizes, move rambling text to the end.</p></li>
<li><p>Add vignette for <code>clusterProfiler</code> and <code>msigdbr</code>.</p></li>
<li>
<p>IN PROGRESS. Convert to Mem S4 object</p>
<ul><li>Update all functions to accept ‘mem’ or ‘Mem’ input.</li>
<li>Update <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> to return ‘Mem’. Consider removing all sections which create cnet/emap objects.</li>
</ul></li>
<li><p>Add emap to <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>, or at least “make it easy”.</p></li>
<li><p>Add Cnet-Heatmap and supporting functions.</p></li>
<li><p>Add testthis unit tests for key behaviors.</p></li>
<li>
<p>ShinyCat</p>
<ul><li><p>Fix bug: singlet node is highlighted, nodesets cannot be highlighted.</p></li>
<li><p>Consider <code><a href="https://rdrr.io/pkg/shinydashboard/man/tabBox.html" class="external-link">shinydashboard::tabBox()</a></code> for inputs; or <code><a href="https://ijlyttle.github.io/bsplus/reference/bs_accordion.html" class="external-link">bsplus::bs_accordion()</a></code></p></li>
<li>
<p>Consider adding global layout options</p>
<ul><li>Relayout with new repulse threshold. Should “nullify” other adjustments.</li>
<li>Nodeset spacing overall. (Rather than having to click each nodeset)</li>
</ul></li>
<li><p>DONE. Consider highlighting edges around the highlighted node/nodeset</p></li>
<li><p>DONE. Add option to save igraph object to a local file</p></li>
<li><p>DONE (removed). Add (or remove) data table tab</p></li>
<li><p>DONE. Document how to retain data</p></li>
</ul></li>
<li>
<p>igraph shapes</p>
<ul><li>clean up use of pie.border, pie.lwd, pie.width, frame.color, frame.width, frame.lwd Official igraph now uses “width” instead of “lwd” so probably best.</li>
<li>Consider JamPolygon? Would also require <code>grid</code>, which would require porting all <code>igraph</code> shapes to <code>grid</code> equivalents. Competes with long-term goal to integrate with ggplot2/tidygraph/ggraph.</li>
</ul></li>
<li>
<p>tidygraph/ggraph long-term</p>
<ul><li>ggraph requires porting ‘jampie’ (and potentially ‘coloredrectangle’) vertex shapes into ggplot2/ggraph.</li>
<li>ggraph may require porting edge bundling to ggraph.</li>
<li>ggraph may require porting ‘qfr’ layout to include argument ‘repulse’.</li>
<li>tidygraph workflows may warrant creating examples which use tidy-oriented logic instead of base R (igraph-like) syntax.</li>
</ul></li>
<li>
<p>reduce R dependencies</p>
<ul><li>DONE. alphahull - requires sp, ggplot2, spatstat. Port <code><a href="https://rdrr.io/pkg/alphahull/man/ashape.html" class="external-link">alphahull::ashape()</a></code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_09oct2025">09oct2025<a class="anchor" aria-label="anchor" href="#id_09oct2025"></a></h2>
<ul><li>
<p>New function: layout with fixed nodes</p>
<ul><li>Purpose is to “make it easy” to apply relayout while holding a subset of nodes fixed in place. Simple wrapper around existing capability, similar to <code><a href="reference/adjust_cnet_set_relayout_gene.html">adjust_cnet_set_relayout_gene()</a></code>. Potential to relayout one nodeset, keeping all others static.</li>
</ul></li>
<li>
<p>DONE. Support “script-like” Cnet adjustments.</p>
<ul><li>Simple text line: node/nodeset<span class="emoji" data-emoji="x">❌</span>y:spacing:rotate</li>
</ul></li>
<li>
<p>Performance:</p>
<ul><li>Response drawing a plot is noticeably slow in R-shiny context, some basic strategies of speeding the process should be applied.</li>
<li>Consider storing “edge bundling group” in <code>igraph</code> edge attributes, since an edge can only belong to one “bundle”. For Cnet, the default is the Cnet nodeset, however any set of edges can be bundled.</li>
<li>Consider pre-calculating edge bunding splines, then caching in <code>igraph</code> attributes for re-use. It must be invalidated by: new layout; any change to layout coordinates. Rotating the igraph layout could also rotate the bundled splines.</li>
<li>Consider pre-calculating mark group alpha hull polygons.</li>
</ul></li>
<li>
<p>DONE. R-shiny Cnet Adjustment Tool ‘ShinyCat’ prototype</p>
<ul><li>Use <code>jam_graph()</code>, associate mouse click x,y to specific node</li>
</ul></li>
<li>
<p>DONE. <code>summarize_igraph_spacing()</code></p>
<ul><li>Defaults: scaled, within-nodeset, each group</li>
<li>call <code><a href="reference/get_cnet_nodeset.html">get_cnet_nodeset()</a></code> when not supplied</li>
</ul></li>
<li>
<p>Improve igraph node label placement:</p>
<ul><li>Consider adding <code>label.adjx</code>, <code>label.adjy</code> to nudge a label, to be applied after <code>label.dist</code> and <code>label.degrees</code>. Unclear if units should be “relative” (by layout) or “absolute”, “relative” seems more convenient as a default.</li>
<li>Reducing label overlaps is a time sink and should be easier, or automated, or both. Related to whether ggraph/tidygraph rendering will ever become viable. If yes then use ggrepel. Using ggraph/tidygraph requires porting <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> features: edge bundling, node shapes (jampie, coloredrectangle). If no, then decide whether to use grid or base R graphics to estimate label sizes, then implement non-overlapping labels (somehow). Ultimately, that process would use <code>label.adjx</code>, <code>label.adjy</code>.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_27mar2025">27mar2025<a class="anchor" aria-label="anchor" href="#id_27mar2025"></a></h2>
<ul><li>Consider approach to store/cache edge bundling for a given layout.</li>
<li>Consider longer-term use of tidygraph/ggraph environment, which requires creating pie node type geom, and foregoing edge bundling until eventually that logic can be ported into that system.</li>
</ul></div>
<div class="section level2">
<h2 id="id_13mar2025">13mar2025<a class="anchor" aria-label="anchor" href="#id_13mar2025"></a></h2>
<ul><li>S4 <code>Mem</code> object should probably also include Mpf output in optional slots.</li>
<li>DONE. Fix: <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> ignores <code>byCols</code> when sorting exemplar pathways.</li>
</ul></div>
<div class="section level2">
<h2 id="id_12feb2025">12feb2025<a class="anchor" aria-label="anchor" href="#id_12feb2025"></a></h2>
<ul><li>
<p><code><a href="reference/importIPAenrichment.html">importIPAenrichment()</a></code></p>
<ul><li>Update to match new variations in the “Export All” text output from IPA.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_15dec2024">15dec2024<a class="anchor" aria-label="anchor" href="#id_15dec2024"></a></h2>
<ul><li>
<p><code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code>, and <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>Consider option to abbreviate pathway names on the fly. Column names are sometimes too ridiculously long, and <code>column_names_max_height</code> isn’t effective enough.</li>
<li>Add the clustering method info to legend as in <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code>.</li>
</ul></li>
<li>
<p>Functions using <code>nodeset</code> names</p>
<ul><li>Protect against sorting errors, e.g. <code>nodeset="D,A"</code> should recognize nodeset whose sorted label is: <code>"A,D"</code>
</li>
<li>See <code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code> and argument <code>set_nodes</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_04dec2024">04dec2024<a class="anchor" aria-label="anchor" href="#id_04dec2024"></a></h2>
<ul><li>
<p>Fix bug “Error in <a href="mailto:gp_hm@column" class="email">gp_hm@column</a>_dend_param$fun(t(<a href="mailto:gp_hm@matrix" class="email">gp_hm@matrix</a>)) : attempt to apply non-function”</p>
<ul><li>Seems caused by small number of pathways in the memIM/enrichIM matrix.</li>
</ul></li>
<li>
<p>Consider removing <code>sf</code>.</p>
<ul><li>Potentially replace with direct calls to <code>polyclip</code>.</li>
<li>Replace with <code>JamPolygon</code>, which would suggest replacing with <code>grid</code>. This would be a “hobby option” - would require updating JamPolygon.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12nov2024">12nov2024<a class="anchor" aria-label="anchor" href="#id_12nov2024"></a></h2>
<ul><li>
<p>Ideas for S4 helper functions / accessors</p>
<ul><li><p><code>exemplars()</code> - extract exemplar pathways for Gene-Pathway clusters</p></li>
<li>
<p><code>as_data_frame()</code> - convert to formal data.frame format</p>
<ul><li><code>Mem</code></li>
<li>
<code>Mpf</code> - one section to describe each pathway cluster</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_02nov2024">02nov2024<a class="anchor" aria-label="anchor" href="#id_02nov2024"></a></h2>
<ul><li>
<p>S4 objects to consider creating:</p>
<ul><li>
<p><code>Mem</code> - MultiEnrichMent</p>
<ul><li>Consider storing MPF clustering params here</li>
<li>Consider storing pathway clusters here, to allow custom assignment of pathways to clusters in mpf</li>
</ul></li>
<li>
<p><code>MPF</code> - Mem Plot Folio</p>
<ul><li>Consider storing <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> output as fixed object. Alternative is to add into <code>Mem</code> above.</li>
</ul></li>
<li>
<p><code>Cnet</code> - Can it inherit from <code>igraph</code>? (I’m really unsure that it would be more benefit than hassle.)</p>
<ul><li>
<p>Methodology</p>
<ul><li>Apparently yes. See <code>setOldClass()</code>; virtual slot <code>.Data</code> for S3 object; <code>as(x, "S3)</code> to convert to S3 form; <code>extends(class(x))</code> as equivalent of S3 inheritance; see <a href="https://stackoverflow.com/questions/64683532/r-as-method-for-converting-custom-s4-class" class="external-link uri">https://stackoverflow.com/questions/64683532/r-as-method-for-converting-custom-s4-class</a>
</li>
</ul></li>
<li>
<p>Benefits</p>
<ul><li>Enforce constraints unique to Cnet-type objects: <code>nodeType</code>, <code>cnet_nodesets</code>, etc.</li>
<li>Direct the <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> function to <code><a href="reference/jam_igraph.html">jam_igraph()</a></code>.</li>
<li>Direct other “cnet” functions to use this object type.</li>
</ul></li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>
<p>Consider adding Mem params to the heatmap legend, e.g.</p>
<ul><li><code>Top Pathways N (topEnrichN)</code></li>
<li><code>Gene min count</code></li>
<li><code>Number of pathway clusters</code></li>
</ul></li>
<li><p>Add Multi-Enrichment Plot to the workflow, <code><a href="reference/mem2emap.html">mem2emap()</a></code>.</p></li>
</ul></li>
<li>
<p>Add <code>mem_cnet_heatmap()</code> - “Cnet-Heatmap”</p>
<ul><li>Cnet network plot (center) surrounded by heatmaps for each Cnet cluster.</li>
</ul></li>
<li>
<p>Consider compatibility with other Bioconductor S4 object classes:</p>
<ul><li>
<code>clusterProfiler::enrichResult</code> - used to some extent already here</li>
<li>
<code>clusterProfiler::gseaResult</code> - not used directly here Ref: <a href="https://doi.org/10.1016/j.xinn.2021.100141" class="external-link uri">https://doi.org/10.1016/j.xinn.2021.100141</a>
</li>
<li>
<code>EGSEA::EGSEAResults</code>, see <code>topSets()</code>, <code>plotHeatmap()</code>, <code>plotPathway()</code>, <code>plotMethods()</code>, <code>plotSummary()</code>, <code>plotGOGraph()</code>, <code>plotBars()</code>, <code>limmaTopTable()</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_01nov2024">01nov2024<a class="anchor" aria-label="anchor" href="#id_01nov2024"></a></h2>
<ul><li>Minor: Consider using <code>cli</code> for messaging, for consistency with tidyverse.</li>
</ul></div>
<div class="section level2">
<h2 id="id_29oct2024">29oct2024<a class="anchor" aria-label="anchor" href="#id_29oct2024"></a></h2>
<ul><li>
<p>General goal is to reduce label overlaps in Cnet plots</p>
<ul><li><p>Many options, including considering <code>ggraph</code> ot <code>tidygraph</code>, however both require re-creating pie/jampie graph node shapes that display both fill colors and border outline (without hiding adjacent borders). These strategies would not respect edge bundling, so that would either be lost or would require porting to ggplot ecosystem somehow. Sigh. Then use <code>ggrepel</code> for non-overlapping labels.</p></li>
<li><p>Ideal world: There exists some <code>grid</code> tool for non-overlapping labels, and not <code>ggplot2</code>.</p></li>
<li>
<p>A simpler option is to allow direct x/y coordinate label adjustment. Current approach requires angle and distance from node center. Difficult to move one label “up and left a tiny amount”. If given ability to adjust a label, it could be possible to adjust all labels to reduce overlaps. Potential workflow:</p>
<ul><li><p>Define label coordinates after calculating angle and distance from node center per usual.</p></li>
<li><p>Apply adjustments.</p></li>
<li><p>Apply user-provided adjustments. Easiest approach. Default nudge matrix is c(0, 0) for all nodes. Edit for specific nodes, relative to plot layout coordinates.</p></li>
<li>
<p>Future: Automated adjustments.</p>
<ul><li>Use <code><a href="https://rdrr.io/r/graphics/strwidth.html" class="external-link">strwidth()</a></code> and <code><a href="https://rdrr.io/r/graphics/strwidth.html" class="external-link">strheight()</a></code> to define bounding boxes.</li>
<li>Adjust bounding boxes?</li>
<li>The <code>plotrix</code> approach performs radial search in local space. It works best when labels are placed in smart order, e.g. center in a cluster, then working out toward edges. Define clusters? Define center of each cluster. Sort labels by distance from cluster center.</li>
<li>Force-directed: calc overlap of two labels, center of overlap, center of bounding box, calculate angle, relative “strength”, and repel along the opposite angle.</li>
</ul></li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_09oct2024">09oct2024<a class="anchor" aria-label="anchor" href="#id_09oct2024"></a></h2>
<ul><li>
<p><code><a href="reference/fixSetLabels.html">fixSetLabels()</a></code></p>
<ul><li>Define better default replacements.</li>
<li>Consider using default from/to, with option for user-defined additions.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_09sep2024">09sep2024<a class="anchor" aria-label="anchor" href="#id_09sep2024"></a></h2>
<ul><li>
<p>Consider using <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> as the top_annotation for <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code>, including for example the dotplot format.</p>
<ul><li>Easiest prototype might be to use <code>+</code> to display the gene-path heatmap rotated with <code>rotate_heatmap=TRUE</code> beside the enrichment heatmap,then merge the color legends:</li>
</ul><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">mpf</span><span class="op">$</span><span class="va">enrichment_hm</span> <span class="op">+</span> <span class="va">mpf</span><span class="op">$</span><span class="va">gp_hm</span>,</span>
<span>   annotation_legend_list<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">mpf</span><span class="op">$</span><span class="va">enrichment_hm</span>, <span class="st">"annotation_legend_list"</span><span class="op">)</span>,</span>
<span>   merge_legends<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</li>
</ul></div>
<div class="section level2">
<h2 id="id_04sep2024">04sep2024<a class="anchor" aria-label="anchor" href="#id_04sep2024"></a></h2>
<ul><li><p>DONE. Fix the <code>row_title</code> for the enrichment heatmap when called by <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>, currently it uses numbers instead of <code>LETTERS</code>.</p></li>
<li><p>DONE. Consider option for fixed-attribute cells for <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> when used with dot plot, so each cell is square with the circle centered.</p></li>
<li>
<p>Consider returning updated <code>igraph</code> from <code><a href="reference/jam_igraph.html">jam_igraph()</a></code></p>
<ul><li>Currently returns <code>invisible(NULL)</code>.</li>
<li>For example, after applying all the updates to <code>igraph</code> values such as node size, label size, label distance, etc.</li>
<li>The end goal is to be able to call <code>jam_igraph(output)</code> with no additional arguments and have it (mostly) render the identical figure.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_30aug2024">30aug2024<a class="anchor" aria-label="anchor" href="#id_30aug2024"></a></h2>
<ul><li>
<p>Add unit tests.</p>
<ul><li>Cover all basic functionality.</li>
<li>Examples and vignettes should already cover the core workflow.</li>
</ul></li>
<li>
<p>Consider adjustments to <code><a href="reference/make_point_hull.html">make_point_hull()</a></code></p>
<ul><li>font size</li>
<li>distance from hull</li>
</ul></li>
<li>
<p>Big picture: Consider using <code>venndir::JamPolygon</code> instead of <code>sf</code> polygons.</p>
<ul><li>Would calculate offsets, rescale, transform, using <code>JamPolygon</code> functions.</li>
<li>Biggest benefit is to reduce dependencies, <code>sf</code> is heavy.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_16aug2024">16aug2024<a class="anchor" aria-label="anchor" href="#id_16aug2024"></a></h2>
<ul><li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>Changed default caption to use <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">ComplexHeatmap::Legend()</a></code> format so it can be included along with color legends.</li>
<li>Increased caption font, it was not legible.</li>
<li>Moved the <code>gene rows, set columns</code> labels to the top, split in two rows.</li>
<li>Return added attribute <code>"caption_legendlist"</code>.</li>
</ul></li>
<li>
<p><code><a href="reference/apply_cnet_direction.html">apply_cnet_direction()</a></code></p>
<ul><li>Change default logic to use <code>pie.border</code> always instead of switching to <code>frame.color</code> when all borders are identical. I think this logic should belong in the plotting function to decide how to render nodes, <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> Also more convenient when reviewing the data, user should only have to look at <code>pie.border</code> and not combination of <code>pie.border</code> and <code>frame.color</code>.</li>
</ul></li>
<li>
<p><code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code></p>
<ul><li>Consider adding <code>top_annotation</code> using the <code>colorV</code> colors per enrichment, for consistency with <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code> and Cnet plots.</li>
</ul></li>
<li>
<p>DONE. Add <code><a href="reference/add_pathway_direction.html">add_pathway_direction()</a></code></p>
<ul><li>DONE. Helper function to add directional z-score column to <code>enrichResult</code> data, using the formula from QIAGEN IPA: <code>z &lt;- (N_genes_up - N_genes_down) / sqrt(N_genes_up + N_genes_down)</code>
</li>
</ul></li>
<li>
<p>Call <code><a href="reference/add_pathway_direction.html">add_pathway_direction()</a></code> from <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> when appropriate.</p>
<ul><li>
<code>geneHitList</code> or <code>geneHitIM</code> are provided, and</li>
<li>contain positive and negative values, and</li>
<li>
<code>"z-score"</code> is not already defined in each <code>enrichResult</code> object. (Bonus points for checking the direction colname attribute.)</li>
</ul></li>
<li>
<p>Review <code>clusterProfiler::compareClusterResult-class</code> object definition. For example <code>clusterProfiler::merge_result(list(enrichResults))</code>.</p>
<ul><li>Consider some form of integration, if possible, for example conversion to/from to call similar functions in <code>clusterProfiler</code>.</li>
</ul></li>
<li>
<p>S4 object <code>Mem</code> to replace <code>list</code> returned by <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li><p>IT IS COMING</p></li>
<li><p>Consider <code><a href="reference/multienrichjam-package.html">multienrichjam()</a></code> to replace <code>multiEnrichmap()</code>?</p></li>
<li>
<p><strong>slots</strong>:</p>
<ul><li>geneIM (im, direction, colors): <code>matrix</code> objects</li>
<li>enrichIM, (pvalues, direction, geneCount, colors): <code>matrix</code> objects</li>
<li>memIM: <code>matrix</code> object</li>
<li>enrichList: <code>list</code> of <code>enrichResult</code> objects</li>
<li>colorV: <code>character</code> vector of colors per enrichment</li>
<li>colnames: <code>character</code> column assignment (consider omitting to enforce standardized colnames)</li>
</ul></li>
<li>
<p><strong>methods</strong>:</p>
<ul><li>
<code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>, supporting functions: <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code>, <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code>, <code><a href="reference/mem2cnet.html">mem2cnet()</a></code>
</li>
<li>
<code>enrichList()</code> - accessor for <code>mem@enrichList</code>
</li>
<li>
<code>mem2dfs()</code> - create series of <code>data.frame</code> summarizing content, intended for export to Excel xlsx.</li>
<li>
<code>mem2xlsx()</code> - direct export to Excel xlsx, calling <code>mem2dfs()</code>.</li>
</ul></li>
<li>
<p><strong>behaviors</strong></p>
<ul><li>
<code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> returns <code>Mem</code> object instead of <code>list</code>
</li>
<li>
<code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> may store parameters in the <code>Mem</code> object</li>
<li>
<code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code>, <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> could also store/retrieve parameters from the <code>mem</code> input object.</li>
<li>
<code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> optionally stores plots into <code>mem</code> to maintian consistent plot attributes</li>
</ul></li>
</ul></li>
<li>
<p>Consider <code>Cnet</code> object that inherits <code>igraph</code>?</p>
<ul><li>It behaves as an <code>igraph</code> object except its class is helpful for generic functions: <code>plot.Cnet()</code>, <code>layout.Cnet()</code>, <code>relayout.Cnet()</code>
</li>
<li>Unclear if S3 object type is preferred since it could inherit <code>igraph</code> (S3 object) characteristics.</li>
</ul></li>
<li>
<p>Consider <code>subgraph.Cnet()</code> or <code>subgraph.igraph()</code> functions</p>
<ul><li>Main purpose is to subset the layout as well as nodes.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_07aug2024">07aug2024<a class="anchor" aria-label="anchor" href="#id_07aug2024"></a></h2>
<ul><li>Remove all <code><a href="https://rdrr.io/r/base/library.html" class="external-link">require()</a></code> checks, since they should already be in Dependencies.</li>
</ul></div>
<div class="section level2">
<h2 id="id_19jul2024">19jul2024<a class="anchor" aria-label="anchor" href="#id_19jul2024"></a></h2>
<ul><li><p>Remove <code>gsubs()</code> which causes a warning upon loading <code>multienrichjam</code>. It conflicts with <code><a href="https://jmw86069.github.io/jamba/reference/gsubs.html" class="external-link">jamba::gsubs()</a></code>. They have slightly different logic.</p></li>
<li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>Enrichment heatmap should define <code>row_title</code> to match <code>pathway_column_title</code>, <code>LETTERS</code> by default. It currently shows numbers.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_11jul2024">11jul2024<a class="anchor" aria-label="anchor" href="#id_11jul2024"></a></h2>
<ul><li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> - option to support RMarkdown output</p>
<ul><li>Provide optional wrapper for RMarkdown output, specifically to print headings/tabs for each plot produced.</li>
<li>Slight downside is there isn’t an easy way to configure a unique figure size for each plot, so plot sizes are at the mercy of the Rmd chunk options <code>fig.height</code>, <code>fig.width</code>.</li>
<li>One implementation option is to allow <code>hook_preplot</code> and <code>hook_postplot</code> to allow the user to run a custom function before and after each plot is drawn. That feels too complicated, when the main driver is just to print an RMarkdown header.</li>
<li>Investigate whether each Markdown tab can define a new figure size.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_21jun2024">21jun2024<a class="anchor" aria-label="anchor" href="#id_21jun2024"></a></h2>
<ul><li>
<p><code><a href="reference/importIPAenrichment.html">importIPAenrichment()</a></code></p>
<ul><li>
<p>DONE. Consider handling gene identifiers so that the default behavior refers to each enriched gene by the original input gene symbol, and not the IPA-generated gene symbol. Some considerations:</p>
<ul><li><p>The main driver is to associate pathway genes to the input data, using the same identifier as the input data. Sometimes IPA assigns its own name which will not match the input data.</p></li>
<li><p>We may consider that data integration (comparison across enrichments) may perform better by comparing via IPA Symbol. Consider the example <code>"HSPA1A/HSPA1B"</code>, one experiment may find <code>"HSPA1A"</code> significant, another may find <code>"HSPA1B"</code> significant. According to IPA, these results are equivalent, in which case the IPA symbol <code>"HSPA1A/HSPA1B"</code> would allow them to use the same identifier.</p></li>
<li><p>Sometimes the “user input” is a platform ID, such as Affymetrix probe. In this case may not be preferable to use the “user input”. In this case it may be convenient for data integration, but less convenient when trying to recognize gene symbols as labels.</p></li>
<li><p>This step requires <code>"Analysis Ready Molecules"</code> is available, and follows expected convention used by Ingenuity.</p></li>
<li>
<p>When multiple genes are combined to “one entity” by IPA, and only one input symbol is retained in the <code>"Analysis Ready Molecules"</code>. The driving use case: <code>"HSPA1A"</code> and <code>"HSPA1B"</code> are combined to one gene entry <code>"HSPA1A/HSPA1B"</code> by IPA. They are considered one gene for the purpose of enrichment. If one or both genes were significant, they would appear as <code>"HSPA1A/HSPA1B"</code> by IPA. The <code>"Analysis Ready Molecules"</code> will list one symbol <code>"HSPA1B"</code> as exemplar, and no entry will appear for <code>"HSPA1A"</code>. In this case there are two options:</p>
<ol style="list-style-type: decimal"><li>Use only the IPA assigned input symbol as provided: <code>"HSPA1B"</code>. IMPLEMENTED with <code>revert_ipa_xref=TRUE</code> (new default).</li>
<li>Split the IPA multi-symbol label into component parts <code>"HSPA1A"</code> and <code>"HSPA1B"</code>. In this case, we need the user to supply the actual gene hits, so we retain only the gene hit the user provided. NOT IMPLEMENTED.</li>
<li>Leave the entry as <code>"HSPA1A/HSPA1B"</code>, however this symbol will not match any input gene hit list, or other expression data matrix. IMPLEMENTED with <code>revert_ipa_xref=FALSE</code>.</li>
</ol></li>
</ul></li>
<li><p>Consider retaining the header section with analysis details, at least for text input.</p></li>
</ul></li>
<li>
<p><code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li>Sometimes the gene rows in <code>geneIM</code> do not match gene rows in <code>memIM</code>, causing an error downstream. The problem appears to happen when the gene hit list does not match all entries in <code>memIM</code>, causing <code>geneIM</code> to have fewer rows.</li>
<li>One solution is to reduce <code>memIM</code> - not ideal because we do not want to lose data.</li>
<li>Another option is to expand <code>geneIM</code> - which requires inferring the incidence matrix. For directional data, it would impose <code>1</code> regardless of the intended directionality, since no other source is available.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_16may2024">16may2024<a class="anchor" aria-label="anchor" href="#id_16may2024"></a></h2>
<ul><li>
<p>edge bundling</p>
<ul><li>Between two communities currently calculates the “central path” between the group centroids. Consider calculating the “central path” between the points with edges between the two communities.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10may2024">10may2024<a class="anchor" aria-label="anchor" href="#id_10may2024"></a></h2>
<ul><li>
<p><code>igraph</code> adjustment scripting language?</p>
<ul><li>
<p>Combines:</p>
<ul><li>
<code>nudge_igraph_nodes()</code>: Fuscobac:<span class="emoji" data-emoji="x">❌</span>0.01:y:-0.02</li>
<li>
<code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code>: nodeset:A::degrees:45:x0.1:y-0.05:percent_spacing:7</li>
<li>
<code><a href="reference/apply_nodeset_spacing.html">apply_nodeset_spacing()</a></code>: nodeset:B::percent_spacing:7</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code>, <code><a href="reference/apply_nodeset_spacing.html">apply_nodeset_spacing()</a></code></p>
<ul><li>Debug issue where <code>label.dist</code> and <code>label.degrees</code> are defined for subset of nodes, leaving other node attributes <code>NA</code> - which causes an error in <code><a href="reference/jam_igraph.html">jam_igraph()</a></code>. Occurs when only one nodeset is adjusted.</li>
<li>Debug issue where <code>percent_spacing</code> fails in <code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code> when supplying custom nodegroups from community detection, and are not proper Cnet nodesets.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_03may2024">03may2024<a class="anchor" aria-label="anchor" href="#id_03may2024"></a></h2>
<ul><li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code> - lower priority but could be useful</p>
<ul><li><p>Consider interactive view (plotly? InteractiveComplexHeatmap?) to enable hover text with the enrichment P-value, gene count, z-score. Could be particularly useful with directional output.</p></li>
<li>
<p>Consider optionally labeling the significant dots?</p>
<ul><li>P-value</li>
<li>z-score</li>
<li>number of genes.</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/mem_legend.html">mem_legend()</a></code> - lower priority, eventually necessary</p>
<ul><li>Consider using <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">ComplexHeatmap::Legend()</a></code> for consistency with other legends, and to allow combining multiple legends together.</li>
</ul></li>
<li>
<p>S4 object <code>mem</code> (or <code>MEM</code>?) - higher priority - sooner the better</p>
<ul><li>
<p>streamlined data content:</p>
<ul><li><p>geneIM (im, direction, colors): <code>matrix</code> objects</p></li>
<li><p>enrichIM, (pvalues, direction, geneCount, colors): <code>matrix</code> objects</p></li>
<li><p>memIM: <code>matrix</code> object</p></li>
<li><p>enrichList: <code>list</code> of <code>enrichResult</code> objects</p></li>
<li><p>colorV: <code>character</code> vector of colors per enrichment</p></li>
<li><p>colnames: <code>character</code> column assignment (consider omitting to enforce standardized colnames)</p></li>
<li><p>omit: <code>multiEnrichMap</code> - in favor of <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>, <code><a href="reference/mem2cnet.html">memIM2cnet()</a></code></p></li>
<li><p>omit: <code>multiCnetPlot</code> - in favor of <code><a href="reference/mem_multienrichplot.html">mem_multienrichplot()</a></code></p></li>
<li><p>optional: store output from <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> to keep a series of plots coordinated, using the same options: <code>pathway_column_split</code>, <code>gene_row_split</code>, <code>enrich_im_weight</code>, <code>gene_im_weight</code>, <code>column_method</code>, <code>row_method</code> (rename <code>column_method</code> to <code>pathway_method</code>?)</p></li>
<li><p>optional <code>enrichment_hm</code>: <code>Heatmap</code> output from <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code>?</p></li>
</ul></li>
<li>
<p>methods:</p>
<ul><li>
<code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>, supporting functions: <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code>, <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code>, <code><a href="reference/mem2cnet.html">mem2cnet()</a></code>
</li>
<li>
<code>enrichList()</code> - accessor for <code>mem@enrichList</code>
</li>
<li>
<code>mem2dfs()</code> - create series of <code>data.frame</code> summarizing content, intended for export to Excel xlsx.</li>
<li>
<code>mem2xlsx()</code> - direct export to Excel xlsx, calling <code>mem2dfs()</code>.</li>
</ul></li>
<li>
<p>behaviors</p>
<ul><li>
<code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> creates <code>MEM</code> object instead of <code>list</code> by default</li>
<li>
<code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code>, <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> could also store/retrieve parameters from the <code>mem</code> input object.</li>
<li>
<code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> optionally stores plots into <code>mem</code> to maintian consistent plot attributes</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li><p>argument <code>do_which</code> - consider accepting <code>character</code> string terms, e.g. <code>"enrichment_hm"</code>, <code>"gp_hm"</code>, <code>"cnet_collapse_set"</code></p></li>
<li>
<p>Consider new argument <code>clusters_mem</code> for these uses:</p>
<ul><li>allow user-defined pathway clusters</li>
<li>allow user-defined pathway subsets (missing pathways are dropped)</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/collapse_mem_clusters.html">collapse_mem_clusters()</a></code></p>
<ul><li>When provided <code>mpf$clusters_mem</code> as <code>list</code> it may result in singlet genes not connected to any pathways - these should (by default) be removed.</li>
</ul></li>
<li>
<p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code> - debug</p>
<ul><li>Apparently sometimes with singlet gene nodes it produces an error: <code>"Error in FUN(X[[i]], ...) : !anyNA(x) is not TRUE"</code> traceback pointed to this line: <code>sf::st_polygon(list(polym)) at jamenrich-igraphshapes.R#1635</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12mar2024">12mar2024<a class="anchor" aria-label="anchor" href="#id_12mar2024"></a></h2>
<ul><li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>
<p>Consider option to place the caption elsewhere.</p>
<ul><li>Problem: The caption sometimes covers part of the color legend in the bottom-right corner.</li>
<li>Another workaround might be to customize the legend layout, so it is not blocked by the caption.</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>, <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>
<p>Consider a workflow to merge pathway clusters, to allow flexibility in how pathway clusters are defined.</p>
<ul><li>Problem: Pathway clusters are sometimes defined inconsistently, where clusters <code>"A"</code> and <code>"B"</code> might be nearly identical.</li>
<li>Problem: It might be visually apparent how to sub-divide pathways, the user may need a mechanism to define pathways to specific clusters.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_04mar2024">04mar2024<a class="anchor" aria-label="anchor" href="#id_04mar2024"></a></h2>
<ul><li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>consider adding column <code>top_annotation</code> with pathway directionality, equivalent to the <code>left_annotation</code> used for gene directionality.</li>
<li>Add row and column annotation padding by default, to help distinguish the central heatmap from the left and top annotations.</li>
</ul></li>
<li>
<p><code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li>when supplied with <code>geneHitIM</code> or <code>geneHitList</code>, calculate the <code>z-score</code> using the IPA formula: <code>z &lt;- (N_genes_up - N_genes_down) / sqrt(N_genes_up + N_genes_down)</code> as described [<a href="https://doi.org/10.1093/bioinformatics/btt703" class="external-link uri">https://doi.org/10.1093/bioinformatics/btt703</a>], and in their FAQ: <a href="https://qiagen.my.salesforce-sites.com/KnowledgeBase/KnowledgeNavigatorPage?id=kA41i000000L5nQCAS&amp;categoryName=BioX" class="external-link">IPA FAQ - Statistical Calculations</a>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_30jan2024">30jan2024<a class="anchor" aria-label="anchor" href="#id_30jan2024"></a></h2>
<ul><li>
<p>DONE. <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li>DONE. <code>geneHitIM</code> and <code>geneHitList</code> are not behaving as intended, nor consistently. They should be interchangeable and equivalent. When <code>geneHitIM</code> is supplied, it populates <code>geneIMdirection</code> but not <code>geneIM</code>. When <code>geneHitList</code> is supplied, it populates <code>geneIM</code> but not <code>geneIMdirection</code>. When either are supplied, they should populate the relevant row in <code>geneIM</code> and <code>geneIMdirection</code>.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_19jan2024">19jan2024<a class="anchor" aria-label="anchor" href="#id_19jan2024"></a></h2>
<ul><li><p>Currently it is cumbersome to edit pathway labels. It could be done for the <code>mem</code> object itself, however the adjustment might need to be different for different plot outputs: heatmap may not work well using word-wrap, while Cnet plots might work best with word-wrap. Publication figures might need an abbreviated label to save plot space.</p></li>
<li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>Consider argument to enable custom adjustment of pathway labels.</li>
</ul></li>
<li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>DONE. Consider argument <code>gene_annotations</code> to enable <code>"geneIM"</code>, <code>"geneIMdirection"</code>.</li>
<li>DONE. When <code>mem$geneIMdirection</code> is present, include directionality in the gene clustering step.</li>
<li>REWRITTEN ABOVE. Consider option to display pathway z-score (<code>mem$pathwayIMdirection</code>) similar to display of <code>mem$geneIMdirection</code>. New argument <code>pathway_annotations</code>.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_06nov2023">06nov2023<a class="anchor" aria-label="anchor" href="#id_06nov2023"></a></h2>
<ul><li><p>DONE. Debug issue when rendering edge arrow heads, they look wonky.</p></li>
<li>
<p>Remove dependency on <code>sf</code> in <code><a href="reference/adjust_polygon_border.html">adjust_polygon_border()</a></code></p>
<ul><li>use <code>polyclip</code> package as it is much more lightweight, without requiring RGEOS, LWGEOM, other world globe map-coordinate based libraries which are not easily compiled on all computer systems using R.</li>
</ul></li>
<li><p>Consider porting <code><a href="reference/deconcat_df2.html">deconcat_df2()</a></code> to <code>jamba</code> for wider re-use by jam packages that would not otherwise need dependency on <code>multienrichjam</code>.</p></li>
<li>
<p>Consider vectorizing edge arrow size</p>
<ul><li>Currently all edge arrows must be the same size (the same limitation is present with <code>igraph::plot()</code>).</li>
<li>This is a niche feature, arrows are rarely used in <code>multienrichjam</code>, and use of differently-sized arrows does not have a driving use case. (This feature is unlikely to move forward until needed.)</li>
</ul></li>
<li>
<p>Consider <code>ggraph</code> compatibility in future.</p>
<ul><li>Many R packages use <code>ggraph</code> for <code>igraph</code> plotting, so it might make sense for consistency to offer this feature. The output <code>ggplot</code> objects can be combined into larger figures using <code>patchwork</code> or <code>cowplot</code> easier than using base R <code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> functions.</li>
<li>My preference not to use <code>ggraph</code> is mainly because the returned <code>ggplot</code> object is not a useful network graph, it is only the instructions for visualization. As such, the layout, node customization, is not persistent outside the <code>ggplot</code> object. The package <code>clusterProfiler</code> migrated to <code>ggraph</code> and now all the returned objects are not useful because they cannot (easily) be customized and analyzed.</li>
<li>However <code>ggraph</code> does not offer <code>pie</code> nor <code>coloredrectangle</code> node shapes.</li>
<li>Note that <code>pie</code> nodes can be emulated with <code>geom_pie()</code> but</li>
</ul><ol style="list-style-type: decimal"><li>is painfully slow to render because it is not vectorized,</li>
<li>does not use inner borders, which allow adjacent wedge borders to be shown beside each other without overlapping.</li>
</ol><ul><li>Task is to add new <code>geom</code> for <code>jampie</code> node shape that accepts pie wedge border colors drawn as inner border, with optional outer border color that also does not overlap the optional inner border.</li>
<li>Another task is to implement edge bundling using <code>ggraph</code> compatible methods. It involves calculating edge bundles, then rendering the curved edges, optionally with arrow heads.</li>
<li>Ultimately a utility function <code>jam_ggraph()</code> may be the <code>ggraph</code> equivalent of <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> for plotting <code>igraph</code> objects.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12oct2023">12oct2023<a class="anchor" aria-label="anchor" href="#id_12oct2023"></a></h2>
<ul><li>
<p>Integrate directionality with more steps in the workflow:</p>
<ul><li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>Consider option to display direction of change in row (gene) annotations for each enrichment. Could display with or without enrichment-colored results?</li>
</ul></li>
</ul></li>
<li>
<p>Create <code>bookdown</code> documentation</p>
<ul><li>Should it use a separate Github repository? (Probably yes.) It looks like <code>"jokergoo/circlize_book"</code> is the repo for the bookdown site.</li>
</ul></li>
<li>
<p>Refactor <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> - maybe new function <code><a href="reference/multienrichjam-package.html">multienrichjam()</a></code>?</p>
<ul><li>
<p>create <code>mem</code> S4 object with slots, print, summary functions</p>
<ul><li>
<p>suggested methods:</p>
<ul><li>
<code><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot()</a></code> could default to <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code> to show all data</li>
<li>
<code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> could print summary of content: enrichments, pathways, genes</li>
<li>consider <code><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame()</a></code> - convert to wide <code>data.frame</code> summary?</li>
<li>
<code>memIM()</code>, <code>geneIM()</code>, <code>enrichIM()</code> convenient access to slot data</li>
<li>convenient way to get <code>list</code> format for IMs?</li>
</ul></li>
<li>
<p>slots:</p>
<ul><li>
<code>memIM</code>: gene/pathway matrix</li>
<li>
<code>geneIM</code>: gene/enrichment matrix</li>
<li>
<code>enrichIM</code>: pathway/enrichment</li>
<li>
<code>geneIMdirection</code>: optional direction per gene/enrichment</li>
<li>
<code>geneIMcolors</code>: colors assigned per gene/enrichment</li>
<li>
<code>enrichIMdirection</code>: optional direction (z-score) per pathway/enrichment</li>
<li>
<code>enrichIMcolors</code>: colors assigned per pathway/enrichment</li>
<li>
<code>enrichIMgeneCount</code>: integer number of genes per pathway/enrichment</li>
</ul></li>
</ul></li>
<li>
<p>remove steps that create embedded Cnet and Emap <code>igraph</code> objects:</p>
<ul><li>
<code>multiCnetPlot</code>, <code>multiCnetPlot1</code>, <code>multiCnetPlot1b</code>, <code>multiCnetPlot2</code>
</li>
<li>
<code>multiEnrichMap</code>, <code>multiEnrichMap2</code>
</li>
<li>
<code>multiEnrichDF</code> - consider saving <code>data.frame</code> with clear name</li>
<li>
<code>multiEnrichResult</code> - What content is stored here?</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_12jul2023">12jul2023<a class="anchor" aria-label="anchor" href="#id_12jul2023"></a></h2>
<ul><li>
<p>New object classes:</p>
<ul><li>
<p><code>"mem"</code>: store output from <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li><p>(Essentially a formal replacement for <code>list</code> format used currently.)</p></li>
<li><p>memIM</p></li>
<li><p>enrichIM, enrichIMcolors, enrichIMdirection</p></li>
<li><p>geneIM, geneIMcolors, geneIMdirection</p></li>
<li><p>geneHitList</p></li>
<li><p>colorV</p></li>
<li>
<p>params:</p>
<ul><li>p_cutoff (from argument <code>cutoffRowMinP</code>)</li>
<li>min_count</li>
<li>topEnrichN</li>
<li>pvalueColname</li>
<li>directionColname</li>
</ul></li>
</ul></li>
<li><p><code>"mem_plots"</code>: store <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> data for re-use.</p></li>
</ul></li>
<li>
<p>New wrapper function <code>multienrich()</code> to replace <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>?</p>
<ul><li>streamlined refactor and replacement for <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>
</li>
<li>avoids defining Cnet and Emap data, pushing into <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>
</li>
<li>reduces arguments by removing all visualization arguments</li>
<li>consider storing <code>memIM</code> data as <code>SummarizedExperiment</code> for convenient use with <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code> via <code><a href="https://rdrr.io/pkg/jamses/man/heatmap_se.html" class="external-link">jamses::heatmap_se()</a></code>.</li>
<li>returns <code>mem</code> object.</li>
</ul></li>
<li>
<p>Update <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li><p>input <code>mem</code> object</p></li>
<li><p>add <code><a href="reference/mem2emap.html">mem2emap()</a></code> plot output.</p></li>
<li>
<p>consider using <code><a href="https://rdrr.io/pkg/jamses/man/heatmap_se.html" class="external-link">jamses::heatmap_se()</a></code> for heatmap functions</p>
<ul><li>it adds <code>jamses</code> as dependency, along with its dependencies</li>
<li>it puts pressure to refactor the jamses contrast stats code</li>
<li>should <code><a href="https://rdrr.io/pkg/jamses/man/heatmap_se.html" class="external-link">jamses::heatmap_se()</a></code> be moved to new package focused only on SummarizedExperiment heatmaps?</li>
<li>See Bioconductor package <code>sechm</code> which is much less capable, but has the same inspiration. It provides <strong>row scaling</strong> (ack) as a recommended (!) option. (Problematic, sorry to say. For gene expression data, the magnitude of change is important, and matters. To rescale the numeric range for <strong>consistency</strong> is counter to the biology, and to the technology. The technology has measurement limitations, for which seeing the <strong>actual</strong> numeric differences is important for assessing whether changes are reliable from that platform. These differences also transfer into follow-up confirmation assays, where changes below a threshold are not feasible to confirm. In general, gene (transcript or protein) expression fold changes are relatively consistently measured for each gene, so to <strong>enhance</strong> the apparent fold change of one gene to fit the fold change of another gene is not necessary, certainly not by default.)</li>
</ul></li>
</ul></li>
<li>
<p>Consider new R package for SummarizedExperiment heatmaps</p>
<ul><li>move <code><a href="https://rdrr.io/pkg/jamses/man/heatmap_se.html" class="external-link">jamses::heatmap_se()</a></code> into this package</li>
<li>move <code><a href="https://rdrr.io/pkg/platjam/man/design2colors.html" class="external-link">platjam::design2colors()</a></code> into this package</li>
<li>minimize R package dependencies</li>
</ul></li>
<li>
<p>R packages to review</p>
<ul><li>Bioconductor package <code>"ConsensusClusterPlus"</code> which can be used to determine appropriate <code>k</code> values for k-means clustering, with metrics to assess consistency of cluster assignment.</li>
<li>
<code>"GeneTonic"</code> function <code>ggs_graph()</code> produces a Cnet plot which they export to <code>visIgraph()</code> for interactivity; <code>enrichment_map()</code> creates EnrichMap.</li>
<li>
<code>"monaLisa"</code> - motif enrichment, extends HOMER theme; nice heatmap Motif labels</li>
<li>
<code>"profileplyr"</code> - coverage heatmap using tidyverse plyr syntax</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_14jun2023">14jun2023<a class="anchor" aria-label="anchor" href="#id_14jun2023"></a></h2>
<ul><li>
<p>DONE: <code><a href="reference/reorderIgraphNodes.html">reorder_igraph_nodes()</a></code></p>
<ul><li>When <code>orderByAspect=TRUE</code>, and it detects tall-skinny aspect ratio, it appears to be applying the y-axis sorting bottom-to-top instead of top-to-bottom.</li>
<li>The culprit was <code><a href="reference/spread_igraph_labels.html">spread_igraph_labels()</a></code> default argument, changed from <code>nodeSortBy=c("x", "y")</code> to <code>nodeSortBy=c("x", "-y")</code>.</li>
</ul></li>
<li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>Slightly increase the spacing between heatmap body and row/column annotations. Currently the gap between heatmap row/column split is identical to the gap between heatmap and row/column annotations, which makes it harder to distinguish one from the other.</li>
<li>The same can be accomplished using <code>ComplexHeatmap::ht_opts()</code> but the option is hard to remember, and would ideally need to be set back to the previous value after drawing the plot.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_05jun2023">05jun2023<a class="anchor" aria-label="anchor" href="#id_05jun2023"></a></h2>
<ul><li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>
<p>DONE: The enrichment dot plot (or enrichment heatmap) should be created after the gene-path heatmap is created, in order to define pathway clusters using gene content, rather than defining pathway clusters using the <code>-log10(Pvalue)</code> matrix.</p>
<ul><li>The clusters derived from the P-value matrix are sometimes not very similar to the gene-pathway clustering result.</li>
<li>For this workflow, it makes the most sense to define pathway clusters upfront, then share those pathway clusters with all downstream visualizations.</li>
<li>Consider creating object class “mem_plots”.</li>
</ul></li>
</ul></li>
<li>
<p>consider new function to convert IPA enrichment data to <code>geneHitList</code></p>
<ul><li>list element <code>"Analysis Ready Molecules"</code> is provided in the IPA output, and this data can be used to re-create the directional hit matrix used during import (if directionality such as fold change was provided to IPA).</li>
</ul></li>
<li>
<p>consider new function to evaluate gene-pathway heatmap output</p>
<ul><li>
<p>The driving use case is selecting <code>pathway_column_split=5</code> upfront, but realizing perhaps 3 clusters would be preferred based upon the content.</p>
<ul><li>Criteria for collapsing two pathway clusters together: either Jaccard similarity above 0.4, or correlation above 0.6.</li>
<li>Definitely requires more testing to determine appropriate default thresholds, or whether a reasonable data-driven threshold can be defined.</li>
</ul></li>
<li>
<p>sometimes two clusters can and perhaps should be merged together</p>
<ul><li>Create collapsed incidence matrix,</li>
<li>Calculate correlation,</li>
<li>Any two clusters with correlation &gt;= 0.2 could be merged?</li>
<li>Alternatively, if more than 3 clusters would be merged, cancel to prevent merging too many clusters together.</li>
</ul></li>
</ul></li>
<li>
<p>consider new function to edit vertex attributes?</p>
<ul><li>
<p>use case: existing <code>igraph</code> object nodes have attributes to modify:</p>
<ul><li>sometimes modify only certain nodes: <code>nodeType="Set"</code> to edit labels</li>
<li>attributes in atomic vector form</li>
<li>attributes (<code>pie.color</code>, <code>pie.border</code>) in <code>list</code> form</li>
<li>also sometimes want to adjust colors - probably separate function</li>
<li>function returns <code>igraph</code> object with attribute modified and stored in the same state as present originally (e.g. atomic remains atomic; list remains list).</li>
</ul></li>
<li>
<p>Which format seems most reasonable?</p>
<ul><li>``` gsub_vertex(g, pattern_l=list( nodeType=c(Set=“^(WP|KEGG|BIOCARTA|GO|REACTOME)_“)), replacement_l=list( nodeType=c(Set=”“)))</li>
</ul><pre><code>
* ```
gsub_vertex(g,
   subset_attr="nodeType",
   subset_attr_value="Set",
   pattern="^(WP|KEGG|BIOCARTA|GO|REACTOME)_",
   replacement="")</code></pre>
<ul><li><pre><code><span> <span class="fu">gsub_vertex</span><span class="op">(</span><span class="va">g</span>,</span>
<span> subset_attr_l<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nodeType<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Set"</span><span class="op">)</span><span class="op">)</span>,</span>
<span> pattern<span class="op">=</span><span class="st">"^(WP|KEGG|BIOCARTA|GO|REACTOME)_"</span>,</span>
<span> replacement<span class="op">=</span><span class="st">""</span><span class="op">)</span></span></code></pre></li>
</ul></li>
</ul></li>
<li>
<p>consider new function to adjust <code>igraph</code> node colors in all forms</p>
<ul><li>modify all fill color attributes <code>color</code>, <code>pie.color</code>, <code>coloredrect.color</code>
</li>
<li>modify all border attributes <code>frame.color</code>, <code>pie.border</code>, <code>coloredrect.border</code>
</li>
<li>adjust <code>frame.width</code>, <code>pie.lwd</code>, <code>coloredrect.lwd</code> relative to each other. For example when <code>pie.lwd</code> and <code>pie.border</code> are both defined (and not transparent), <code>frame.width=0.1</code>, otherwise <code>frame.width=2</code>.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_02jun2023">02jun2023<a class="anchor" aria-label="anchor" href="#id_02jun2023"></a></h2>
<ul><li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>Debug edge cases where <code>pathway_column_split=4</code> does not match the output number of column split following hierarchical clustering.</li>
<li>Debug edge case where there are not enough pathways or genes to support the gene-pathway heatmap workflow. E.g. only 1 or 2 pathways, or only 1 or 2 genes.</li>
</ul></li>
<li>
<p><code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li>
<p>Debug error “multiEnrichMap(): geneHitIM does not contain 5 rows present in geneIM, default values will use 1.”</p>
<ul><li>Apparently when some <code>rownames(mem$geneIM)</code> are not found in <code>rownames(geneHitIM)</code>.</li>
<li>Ideally change <code>geneIMdirection</code> so it does not store “+1” and instead shows zero or NA. Genes should be shown but without associated direction.</li>
<li>The message should describe how to find missing values so the user can debug the error.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_01jun2023">01jun2023<a class="anchor" aria-label="anchor" href="#id_01jun2023"></a></h2>
<ul><li>
<p><code><a href="reference/reorderIgraphNodes.html">reorder_igraph_nodes()</a></code></p>
<ul><li>Currently <code>orderByAspect=TRUE</code> will order nodes based upon the aspect ratio of each nodeset.</li>
<li>Ideal world: when nodes are sorted by something like color, calculate the aspect ratio of nodes within that color. Situation: Assume a tall-skinny nodeset, sorted top-to-bottom by different colors. One color has 12 entries, the nodeset is 12 nodes wide, so this color appears horizontal among the other nodes. When sorting by border color, it goes top-to-bottom, which is not visually intuitive.</li>
</ul></li>
<li>
<p><code><a href="reference/apply_cnet_direction.html">apply_cnet_direction()</a></code></p>
<ul><li>DONE: changed default <code>col</code> to use colors: <code>c("blue", "grey80", "firebrick3")</code> with breaks <code>c(-1, 0, 1)</code>.</li>
</ul></li>
<li>
<p>change all <code>frame.lwd</code> to <code>frame.width</code> before <code>frame.lwd</code> is widely used.</p>
<ul><li>consider backward compatibility: when <code>frame.lwd</code> is defined in an <code>igraph</code> object, copy its values into <code>frame.width</code>, then proceed using <code>frame.width</code> for all other operations.</li>
</ul></li>
<li>
<p><code><a href="reference/mem_legend.html">mem_legend()</a></code></p>
<ul><li>DONE: new argument <code>pt.lwd=2</code> to control the line width used only for point borders, useful when <code>do_direction=TRUE</code>. The argument <code>pt.lwd</code> already gets passed to <code><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend()</a></code> however making it a formal argument here helps make the option more clear for users.</li>
<li>Auto-detect whether to enable <code>do_direction=TRUE</code>, by checking if any <code>frame.color</code> or <code>pie.border</code> are defined with red/blue colors. Bonus points for using the same colors defined in <code>frame.color</code> or <code>pie.border</code>, however that may be risky if those colors vary based upon some fold change value, or vary based upon contrasting with the node or pie fill color.</li>
</ul></li>
<li>
<p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code></p>
<ul><li>
<p>Consider handling <code>V(g)$shape="circle"</code> as shape <code>"jampie"</code> during rendering. Certain older Cnet plot <code>igraph</code> objects appear to break the default <code>shape="circle"</code> rendering, some cryptic error about <code><a href="https://rdrr.io/r/base/names.html" class="external-link">names()</a></code> not being defined when expected. The error does not appear with <code><a href="https://r.igraph.org/reference/plot.igraph.html" class="external-link">igraph::plot.igraph()</a></code>, so it is specific to <code><a href="reference/jam_igraph.html">jam_igraph()</a></code>.</p>
<ul><li>Possible workaround is to sidestep the problem by re-using the same <code>shape="jampie"</code> rendering method already implemented, which would keep all borders consistent when displaying a mixture of <code>shape="circle"</code> and <code>shape="jampie"</code> nodes.</li>
<li>Problem with that workaround, if a node has <code>pie.color</code> defined it will be used when <code>shape="jampie"</code> even if the user specified <code>shape="circle"</code> (which should only use <code>color</code>). In that case, when a <code>shape="circle"</code> node is being rendered internally as <code>shape="jampie"</code> it should first copy <code>color</code> into <code>pie.color</code> for those nodes; <code>frame.color</code> to <code>pie.border</code>; and <code>frame.lwd</code> (<code>frame.width</code>) to <code>pie.lwd</code>.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_24may2023">24may2023<a class="anchor" aria-label="anchor" href="#id_24may2023"></a></h2>
<ul><li>
<p>DONE: Fix bug with node rendering, caused by recent version of <code>igraph</code> adding <code>vertex.frame.width</code> (and not <code>vertex.frame.lwd</code> ugh).</p>
<ul><li>
<code>NULL</code> or missing <code>vertex.frame.width</code> causes an error. Ultimately caused by no default value defined in the custom function <code><a href="reference/default_igraph_values.html">default_igraph_values()</a></code>, which was necessary to create since <code>igraph</code> does not export that function.</li>
<li>FIXED by adding <code>vertex.frame.width</code> to default values.</li>
<li>Longer term fix is to replace all references to <code>vertex.frame.lwd</code> with <code>vertex.frame.width</code>, before the precedent is set.</li>
</ul></li>
<li>
<p>Fix errors caused by <code>"stringsAsFactors=TRUE"</code></p>
<ul><li>DONE: <code><a href="reference/rank_mem_clusters.html">rank_mem_clusters()</a></code>
</li>
</ul></li>
<li>
<p>Fix errors caused when there is only one (or zero) genes.</p>
<ul><li>
<code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code> and <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_20apr2023">20apr2023<a class="anchor" aria-label="anchor" href="#id_20apr2023"></a></h2>
<ul><li>
<p><code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> color legend changes:</p>
<ul><li>Show actual P-value <code>c(1, 0.05, 0.01, 0.001, 10^-4, etc.)</code> using <code>expression</code> for labels, and continue using <code>-log10(p)</code> for color assignment.</li>
<li>Add this argument <code>heatmap_legend_param=list(break_dist=1)</code> which causes the numeric labels to be evenly spaced, instead of having the labels at uneven intervals, often with angled lines connecting to the color legend.</li>
<li>Option for discrete color legend? I.e. Show colors only at the labels, and not show the intervening gradient. It is more difficult to show abrupt transitions, e.g. it would need to show <code>c(1, 0.051, 0.05, 0.01)</code> in order to show that <code>0.051</code> is not colorized, but <code>0.05</code> is colorized. The smooth gradient is actually more effective at conveying that effect without additional labels.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_31mar2023">31mar2023<a class="anchor" aria-label="anchor" href="#id_31mar2023"></a></h2>
<ul><li><p>Nodes with <code>shape="jampie"</code> and <code>frame.lwd=0</code> are still rendering the frame color outside the inner borders. When <code>frame.lwd=0</code> there should be no frame drawn even when <code>frame.color</code> is defined with a color.</p></li>
<li>
<p>Big picture musing: Consider replacing base R plotting functions with corresponding <code>grid</code> functions.</p>
<ul><li>The <code>vwline</code> R package (P. Murrell) is capable of drawing internal/external lines.</li>
<li>The <code>gridGraphics</code> package also provides better methods of clipping curved lines to the edge of a node border for example.</li>
<li>Major downside, it would likely involve rewriting all the <code>igraph</code> node shape functions into corresponding <code>grid</code> format. It is effectively similar to repeating much of <code>ggraph</code>, except that this approach can be customized. The <code>ggraph</code> approach in <code>ggplot2</code> is more or less untouchable in terms of providing customization. Ugh.</li>
<li>Another option may be to figure out how to add custom node shapes to <code>ggraph</code>, then use <code>vwline</code>/<code>gridGraphics</code> for rendering. Also need to write custom edge bundling function, since those in <code>ggraph</code> are wholly insufficient. Yeah, this is a no for now.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_26feb2023">26feb2023<a class="anchor" aria-label="anchor" href="#id_26feb2023"></a></h2>
<ul><li>
<p><code><a href="reference/reorderIgraphNodes.html">reorder_igraph_nodes()</a></code>, <code><a href="reference/reorderIgraphNodes.html">reorderIgraphNodes()</a></code></p>
<ul><li>DONE: method to specify specific nodes or nodesets to be reordered</li>
<li>motivation is to allow sorting based upon relative aspect ratio of nodes, so a nodeset whose nodes are “tall-skinny” can be sorted top-to-bottom, and nodeset whose nodes are “short-wide” can be sorted left-to-right. Frankly, not sure if the inconsistency works for all network layouts, but for sure the top-to-bottom is not ideal for “tall-skinny” nodesets, it is not visually intuitive.</li>
</ul></li>
<li>
<p>consider new igraph shapes, intended to enable inner/outer border, <code>frame.lwd</code></p>
<ul><li>Do these make sense?
<ul><li>
<code>shape.jamcircle.plot()</code> - enable custom <code>frame.lwd</code> for shape=“circle”</li>
<li>others: square, csquare, rectangle, vrectangle</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/label_communities.html">label_communities()</a></code></p>
<ul><li>generalize this method to determine keywords most represented in any set of pathway names.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_02feb2023">02feb2023<a class="anchor" aria-label="anchor" href="#id_02feb2023"></a></h2>
<ul><li>
<p>Low priority visual enhancement, color Cnet edges by Set.</p>
<ul><li>Consider coloring Cnet edges by categorical color, by Set It may help clarify node groupings, while also reinforcing which edges connect to particular nodesets. Unclear whether added color would be confusing or beneficial.</li>
<li>Referring to Neely et al paper in ACR Open Rheumatology: “Gene Expression Meta-Analysis Reveals Concordance in Gene Activation, Pathway and Cell-Type Enrichment in Dermatomyositis (DM) Target Tissues”</li>
<li>They show Cnet-type plot in Figure 4A, where connections from each set were categorically colored</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_26jan2023">26jan2023<a class="anchor" aria-label="anchor" href="#id_26jan2023"></a></h2>
<ul><li>
<p>Low priority: It may be useful to create vectorized functions:</p>
<ul><li>
<p><code><a href="https://rdrr.io/r/graphics/polygon.html" class="external-link">polygon()</a></code></p>
<ul><li>enable multiple line widths for multiple polygons split by <code>NA</code> coordinates.</li>
<li>enable optional inner and outer borders with varying widths</li>
</ul></li>
<li><p><code><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text()</a></code>: enable multiple family, srt</p></li>
<li><p><code><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines()</a></code>: enable multiple col, lty, lwd for split lines, similar to <code><a href="https://rdrr.io/r/graphics/segments.html" class="external-link">segments()</a></code> except enabled for multiple lines split by <code>NA</code> coordinates.</p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_23jan203">23jan203<a class="anchor" aria-label="anchor" href="#id_23jan203"></a></h2>
<ul><li>
<p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code></p>
<ul><li>Return the input <code>igraph</code> object with all relevant object attributes updated to reflect the plot parameters. The returned object could be plotted directly without any customization.</li>
<li>Consider storing/using edge coordinates inside the <code>igraph</code> object. However, whenever layout is re-calculated, edge coordinates would likely become invalid. It would be tricky to handle.</li>
</ul></li>
<li>
<p>edge bundling and edge clipping integration</p>
<ul><li><p>DONE: Most scenarios described below.</p></li>
<li><p>Currently, linear edges are clipped at connected node boundaries.</p></li>
<li><p>When edges are slightly curved, the start and end positions are reasonable.</p></li>
<li><p>When edges are bundled, and especially when nodes are relatively large, the edge curves through the node in a different direction than a linear edge.</p></li>
<li>
<p>This situation is not a visualization problem when:</p>
<ol style="list-style-type: decimal"><li>nodes are not filled with transparent color, and</li>
<li>edges are not drawn with arrows.</li>
</ol></li>
<li>
<p>The situation is only a visible problem when:</p>
<ol style="list-style-type: decimal"><li>edges have arrows that would now be partly or fully covered by the node, or</li>
<li>nodes that are filled with partial or fully transparent color, thus showing the edge underneath.</li>
</ol></li>
<li><p>Proper edge clipping would probably be done by calculating the edge from the node actual center point, then clipping edge where it exits the node shape border.</p></li>
<li>
<p>Implementation may benefit from storing the edge coordinates as an edge attribute, to be used by the clipping function when present. Absence of custom edge coordinates would cause the clipping function to use linear edge coordinates.</p>
<ul><li>The plot function could also use stored edge coordinates as opposed to calling edge bundling function; alternatively the edge bundling function could simply re-use existing edge coordinates as well.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_18jan2023">18jan2023<a class="anchor" aria-label="anchor" href="#id_18jan2023"></a></h2>
<ul><li>
<p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code>, proposed drop-in replacement for <code>igraph:::plot.igraph()</code>:</p>
<ul><li>FIXED: This function does not seem to handle edge arrows, nor does it shorten edges based upon node sizes.</li>
<li>bonus points for node/edge legend functions</li>
<li>When layout is not defined, the xlim/ylim values sometimes do not match the dynamic layout calculated on the fly.</li>
</ul></li>
<li>
<p>igraph shape “ellipse”</p>
<ul><li>DONE: it should have a proper “clip” function, in order for edge arrows to appear at the border of each node</li>
</ul></li>
<li>
<p>igraph “coloredrectangle” shape</p>
<ul><li>DONE: The coloredrect.border should also be capable of adjacent lines that do not overlap.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_30nov2022">30nov2022<a class="anchor" aria-label="anchor" href="#id_30nov2022"></a></h2>
<ul><li>
<p><code><a href="reference/reorderIgraphNodes.html">reorder_igraph_nodes()</a></code></p>
<ul><li>DONE: to be fancy, it should also propagate changes to <code>label.dist</code> and <code>label.degree</code>, as created by <code><a href="reference/spread_igraph_labels.html">spread_igraph_labels()</a></code>, since switching coordinates for two nodes should also switch the <code>label.degree</code> and <code>label.dist</code> associated with those coordinates.</li>
</ul></li>
<li>
<p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code></p>
<ul><li>It shows some lag before plotting all nodes vectorized, check if the edge bundling is the slow step, and optimize as needed. UPDATE: Yes the edge bundling step is introducing lag; or the resulting curved lines are being plotted slowly? Another good reason to store edge bundle coordinates in the <code>igraph</code>.</li>
</ul></li>
<li>
<p><code><a href="reference/mem_legend.html">mem_legend()</a></code></p>
<ul><li>consider using <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Legend.html" class="external-link">ComplexHeatmap::Legend()</a></code> for consistency with future Cnet-Heatmap usage, and because that mechanism is really nice.</li>
</ul></li>
<li>
<p>Refactor <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li><p>likely create new function <code>multienrich()</code> so <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> can be deprecated and remain for backward compatibility.</p></li>
<li><p>Avoid pre-calculating Cnet and Emap graphs. Counterpoint: It could call <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> with defaults, and store results back into the <code>multienrichResult</code>.</p></li>
<li>
<p>Consider <code>multienrichResult</code> object type?</p>
<ul><li>could hold what is now a <code>list</code> object, with proper slot names.</li>
<li>custom <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code>/<code><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary()</a></code> function to display summary info about number of genes, pathways, etc.</li>
</ul></li>
<li>
<p>Consider <code>multienrichPlot</code> object type? No, the decision is to re-use <code>multienrichResult</code>.</p>
<ul><li>Benefit is simplicity: this object feeds all downstream plots.</li>
<li>The negative is that it requires saving a separate <code>multienrichResult</code> with alternative clustering if necessary.</li>
<li>Main goal is to define the gene-pathway clustering, then re-use clusters in subsequent steps without having to repeat the clustering the same way. If one uses custom clustering in <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> to produce gene-pathway heatmap, they have to use the same arguments in subsequent steps otherwise the clusters will differ, and it is not clearly indicated to be a problem.</li>
</ul></li>
<li><p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> may likely return the <code>multienrichResult</code> after updating internally stored data. It should have option to use existing results.</p></li>
</ul></li>
<li>
<p>Write a more directed vignette showing at least two common use cases:</p>
<ol style="list-style-type: decimal"><li>Enrichment using <code><a href="https://rdrr.io/pkg/clusterProfiler/man/enricher.html" class="external-link">clusterProfiler::enricher()</a></code>, then multi-enrichment.</li>
<li>Enrichment using Qiagen Ingenuity IPA (outside of R) then importing the files produced using <code>"Export All"</code> from within the Ingenuity IPA app.</li>
<li>Enrichment using some other external (non-R) tool, for example DAVID.</li>
</ol></li>
<li>
<p>Write a vignette focused solely on Cnet plot custom layout options. Background:</p>
<ul><li><p>Very common workflow results in Cnet plot to summarize the findings.</p></li>
<li><p>This Cnet plot has often been included as a figure or supplementary figure for a paper, therefore it requires manual adjustments to increase legibility and clarity of the figure.</p></li>
<li><p>Adjust individual nodes: <code><a href="reference/nudge_igraph_node.html">nudge_igraph_node()</a></code>. Note this function can be called on a <code>list</code> of nodes using <code>nodes_xy</code>, or using vectors with <code>nodes</code>, <code>x</code>, <code>y</code>.</p></li>
<li>
<p>Adjust sets of nodes:</p>
<ul><li>
<code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code> - Usually for sub-clusters of gene nodes, move the whole group, adjust intra-group node spacing, or rotate the group around the group center.</li>
<li>
<code><a href="reference/apply_nodeset_spacing.html">apply_nodeset_spacing()</a></code> - Apply a minimum spacing between nodes, for each nodeset.</li>
<li>
<code><a href="reference/adjust_cnet_set_relayout_gene.html">adjust_cnet_set_relayout_gene()</a></code> - Adjust “Set” nodes manually, then fix them in place but allow “Gene” nodes to move during re-layout, usually with <code><a href="reference/relayout_with_qfr.html">relayout_with_qfr()</a></code>
</li>
</ul></li>
<li>
<p>Adjust all nodes:</p>
<ul><li>
<code><a href="reference/layout_with_qfr.html">layout_with_qfr()</a></code>, <code><a href="reference/layout_with_qfrf.html">layout_with_qfrf()</a></code>, <code><a href="reference/relayout_with_qfr.html">relayout_with_qfr()</a></code> - All are wrappers to <code><a href="https://rdrr.io/pkg/qgraph/man/qgraph.layout.fruchtermanreingold.html" class="external-link">qgraph::qgraph.layout.fruchtermanreingold()</a></code> with convenient defaults. <code><a href="reference/layout_with_qfr.html">layout_with_qfr()</a></code> returns coordinates; <code><a href="reference/layout_with_qfrf.html">layout_with_qfrf()</a></code> returns a layout function with user-defined <code>repulse</code> argument; <code><a href="reference/relayout_with_qfr.html">relayout_with_qfr()</a></code> updates layout in-place for an <code>igraph</code> object, storing in <code>graph_attr(g, "layout")</code>.</li>
<li>
<code><a href="reference/rotate_igraph_layout.html">rotate_igraph_layout()</a></code> - rotate the layout coordinates by some user-defined degree angle.</li>
<li>
<code><a href="reference/spread_igraph_labels.html">spread_igraph_labels()</a></code> - positions node labels radially around nodes, based upon the average incoming edge angle.</li>
<li>
<code><a href="reference/reorderIgraphNodes.html">reorder_igraph_nodes()</a></code> - within each nodeset, reposition nodes in order of node color, border color, then label. A nodeset is defined as a set of nodes that all share the same connections, which is mostly only useful for bipartite graphs such as Cnet plots. This function performs the node re-ordering for all nodesets, across the whole <code>igraph</code> object.</li>
</ul></li>
<li>
<p>plot with <code><a href="reference/jam_igraph.html">jam_igraph()</a></code></p>
<ul><li>Vectorized plotting when multiple shapes are used, otherwise <code>igraph::plot()</code> uses a <code>for()</code> to iterate each node.</li>
<li>Improved <code>pie</code> rendering, also vectorized.</li>
<li>Convenience methods to adjust node size, node label font size, node label distance</li>
<li>Optional shadow text node labels</li>
<li>Maintain aspect ratio = 1, so nodes are symmetrically spaced along each axis (defined by the layout algorithm used.)</li>
<li>optionally render node groups using <code>mark.groups</code>
</li>
<li>Call edge bundling, especially useful for bipartite graphs such as Cnet plots.</li>
<li>Optionally draw background grid with percent layout units.</li>
</ul></li>
</ul></li>
<li><p>add dev functions <code>layout_cnet()</code> and related iterative layout functions.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_23nov2022">23nov2022<a class="anchor" aria-label="anchor" href="#id_23nov2022"></a></h2>
<ul><li>
<p>mechanism to store edge coordinates in <code>igraph</code> object</p>
<ul><li><p>to my knowledge, this functionality does not exist in <code>igraph</code>, nor is it represented in <code>ggraph</code> or <code>tidygraph</code> objects. However <code>tidygraph</code> may have capability to supply specific edge coordinates if they exist, so it might be the closest to implementing this feature.</p></li>
<li><p><code>igraph::plot()</code> is not equipped to use edge coordinates</p></li>
<li><p><code>ggraph</code> is not equipped to use edge coordinates, it only creates edge coordinates based upon the edge <code>geom_</code> being used.</p></li>
<li>
<p>The driving use case is to define edge coordinates to handle:</p>
<ol style="list-style-type: decimal"><li>edge bundling, a procedure that could be done dynamically, but is computationally expensive for large numbers of nodes;</li>
<li>custom edge pathing, specifically for pathway schematics, such as those generated when using GraphViz DOT format. While the DOT format generates and stores edge coordinates, I could not find examples in R that import a fully-described DOT file (with edge coordinates embedded) that also imported edge coordinates.</li>
</ol></li>
<li>
<p>Reasons to want edge coordinates upfront:</p>
<ul><li>pre-calculate edge bundling, saving time during rendering</li>
<li>allow custom definition of edges, for example in a pathway schematic where edges are positioned to avoid overlaps.</li>
<li>calculate better label placement by using the angle of incoming edges, not limited to the linear vector from node1 to node2.</li>
</ul></li>
<li>
<p>Issues raised when storing edge coordinates:</p>
<ul><li>Obviously, whenever the node layout coordinates are changed, the edges also need to be changed.</li>
<li>
<code><a href="reference/rotate_igraph_layout.html">rotate_igraph_layout()</a></code> could rotate nodes and all edges together.</li>
<li>
<code><a href="reference/nudge_igraph_node.html">nudge_igraph_node()</a></code> and <code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code> must decide whether to adjust edges by simple scaling, or simply invalidate all edges, then force the user to re-calculate edge coordinates.</li>
</ul></li>
<li>
<p>useful helper functions</p>
<ul><li>
<code>validate_edge_coords()</code> - Test whether edge coordinates match node coordinates. If not, then delete or replace edges with linear equivalent.</li>
<li>
<code>adjust_edge()</code> - Wrapper function to adjust the curvature, placement, rotation, of edges. Could be called when rotating node layout, to rotate edges accordingly.</li>
<li>
<code>bundle_edges()</code> - Wrapper to apply bundling to one set of edges together as a group. Optionally define specific coordinate(s) through which the bundling loess curve is routed.</li>
<li>fancy options like routing edges with preference for vertical/horizontal pathing, with slight curvature at right angle turns. Often used in schematic diagrams.</li>
<li>fancy “subway” style options, where bundled edges are allowed to remain visible adjacent to other edges along their path. Edges could “snap” to nearby edges in the same bundle.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_17nov2022">17nov2022<a class="anchor" aria-label="anchor" href="#id_17nov2022"></a></h2>
<ul><li>
<p><code><a href="reference/get_cnet_nodeset.html">get_cnet_nodeset()</a></code></p>
<ul><li>This function is called several times by internal functions, and could therefore be much faster than currently.</li>
<li>Refactor by using <code><a href="https://r.igraph.org/reference/as_adjacency_matrix.html" class="external-link">igraph::as_adjacency_matrix()</a></code>. Subset rows for <code>nodeType="Gene"</code> and columns for <code>nodeType="Set"</code>. Then should be able to convert rapidly by <code><a href="https://jmw86069.github.io/venndir/reference/im2list.html" class="external-link">venndir::im2list()</a></code> or some <code>pasteByRow()</code> magic, produce nodeset per node (row). Finally, split node names by nodeset.</li>
</ul></li>
<li>
<p>node adjustment to prevent label overlaps</p>
<ul><li>Idea: “stretch” out nodeset nodes “to the right”, which fixes the left edge of the nodeset, then expands the node spacing outward as fraction of current range of nodeset nodes. For example, expand to the right by 10%, to improve side-to-side spacing, since label overlaps typically occur with nodes at the same y-level.</li>
<li>Stretch nodes “to the left” would work the same, but fixes the right x-coordinate range, and expands the left x-coordinate range.</li>
<li>Stretch nodes “to the top”, and “to the bottom” work similarly.</li>
<li>Future idea to reduce node label overlap is to treat it like biscuit dough under a rolling pin. Stretch subset of nodes in each direction until the labels no longer overlap. The trick is to stretch nodes away from other nodesets, so it does not cause new overlaps with other nodes.</li>
</ul></li>
<li>
<p><code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code></p>
<ul><li>Consider option to restrict “expand” to x- or y-axis expansion. Basic idea is to limit expansion to “widen” the node spacing, or to make node spacing “taller”. The “widen” option is helpful to reduce label overlaps for nodes directly beside each other.</li>
</ul></li>
<li>
<p><code><a href="reference/spread_igraph_labels.html">spread_igraph_labels()</a></code></p>
<ul><li>ideal case: somehow take into account the edge bundling to calculate input angle to each node, rather than straight vector from node to node.</li>
<li>
<code>node_groups</code> - spread labels relative to node group centroid, so labels in this cluster of nodes will be spaced out from each other. Bonus points for taking into account the overall average input angle to nodes in each group, and applying a fraction of that offset along with the node-to-group offset. For example, for a node group in the top right, they generally point to the top-right, but are fanned out slightly so the bottom-left-most node is not fanned out to bottom-left, but maybe center, or only slightly bottom-left of the node.</li>
</ul></li>
<li>
<p>new layout functions specific for Cnet plots</p>
<ul><li><p><code>iterate_qfr_layout()</code> - R code version of <code><a href="https://rdrr.io/pkg/qgraph/man/qgraph.layout.fruchtermanreingold.html" class="external-link">qgraph::qgraph.layout.fruchtermanreingold()</a></code> with custom addition of node “shells”, and option to call <code>iterate_node_group_distance()</code></p></li>
<li><p><code>iterate_node_group_distance()</code> - R layout intended only to enforce separation across node groups (defined by <code><a href="reference/get_cnet_nodeset.html">get_cnet_nodeset()</a></code>) so there is additional space between nodesets in the layout.</p></li>
<li>
<p><code>layout_cnet()</code> - wrapper function that calls rounds of layouts. This series of steps is currently the best default layout to generate the most readable Cnet plot possible.</p>
<ol style="list-style-type: decimal"><li>initial node placement - qfr layout without node/group distance</li>
<li>node spacing - qfr with node distance</li>
<li>node/group spacing - qfr with node and group distance 4a. imposed nodeset percent spacing</li>
<li>group spacing - “fix” node layout per nodeset, then layout nodesets</li>
<li>detect best rotation to place first “Set” node top-left, then rotate</li>
<li>spread labels, re-order nodes by color/border.</li>
</ol></li>
<li>
<p>Other ideas:</p>
<ol style="list-style-type: decimal"><li>Try the new <code>bubble_force()</code> calculation for minimum distance, instead of the linear desired distance force.</li>
<li>Consider constraining all Gene nodes, then allow only Set nodes to move. Goal is to prevent Set nodes from being embedded inside Gene nodes due to overall net forces.</li>
<li>Optional fixed coordinate range, to prevent layouts from becoming infinitely large, thus endless cycle of imposing minimum precent spacing, making the range larger, thus making the spacing smaller, etc. etc. It means when a force would push a node/group out of bounds, it gets stopped at the boundary/boundaries. Hopefully because forces are applied to both node1 and node2, when node1 cannot move, node2 should still move albeit at half speed.</li>
<li>Really pie in the sky: Use the node size and shape to calculate distance between nodes, nearest polygon distance for example. Probably also a processing non-starter since it would add overhead to the calculations, however it is the type of thing that could be parallelized during each iteration. No idea how efficient it would be to split threads during each iteration.</li>
</ol></li>
<li>
<p>Code it in Rust, use R package <code>expandr</code> to integrate the Rust function into the package via an R function. Follow C++ code steps used by <code>qgraph:::qgraph_layout_Cpp()</code>. It doesn’t seem that these layout iterations are particularly good for multi-threading, however.</p>
<ul><li>each iteration could be multi-threaded. Threads could share one distance matrix, the extract values when needed. Or threads could share node coordinates, then calculate distance when needed on the fly. If fast enough, distance matrix (and memory allocation) could be avoided, just do the math when needed.</li>
<li>Look for Rust libraries for geometric calculations.</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>add option to display gene incidence matrix (left annotation) using <code>geneIMdirection</code> colors, to represent up/down.</li>
</ul></li>
<li>
<p>add <code>plot_cnet_heatmaps()</code></p>
<ul><li>make default arguments as minimal and close to default settings as possible</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10nov2022">10nov2022<a class="anchor" aria-label="anchor" href="#id_10nov2022"></a></h2>
<ul><li>
<p><code><a href="reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code></p>
<ul><li>COMPLETE: add minimum percent spacing, similar to <code><a href="reference/apply_nodeset_spacing.html">apply_nodeset_spacing()</a></code> except to operate on only one nodeset at a time.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_04nov2022">04nov2022<a class="anchor" aria-label="anchor" href="#id_04nov2022"></a></h2>
<ul><li>
<p><code><a href="reference/shape.jampie.plot.html">shape.jampie.plot()</a></code> which calls <code><a href="reference/jam_mypie.html">jam_mypie()</a></code>:</p>
<ul><li><p>When drawing borders for multiple pie shapes, the overlapping border are overdrawn, hiding all but the last border drawn.</p></li>
<li><p>Instead, draw border inside the polygon edge, so the borders are adjacent and not overlapping.</p></li>
<li><p>Quick survey revealed no drop-in replacement <code><a href="https://rdrr.io/r/graphics/polygon.html" class="external-link">polygon()</a></code> functions.</p></li>
<li>
<p>Workaround involves <code><a href="https://r-spatial.github.io/sf/reference/geos_unary.html" class="external-link">sf::st_buffer()</a></code> to calculate a buffer inside each polygon, creating a “donut” filled with the border color. In principle, this issue tends to occur in relatively few nodes for typical Cnet plots, however it could be substantial performance hit.</p>
<ol style="list-style-type: decimal"><li>Each polygon, convert to corresponding <code><a href="https://r-spatial.github.io/sf/reference/st.html" class="external-link">sf::st_polygon</a></code> object</li>
<li>Call <code><a href="https://r-spatial.github.io/sf/reference/geos_unary.html" class="external-link">sf::st_buffer()</a></code>.</li>
<li>Create polygon from original border, inner buffer border. Or split the original polygon at this border, apply color to the outer and inner polygons. This way the border and fill colors are drawn together so nodes are “complete”.</li>
</ol></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_26oct2022">26oct2022<a class="anchor" aria-label="anchor" href="#id_26oct2022"></a></h2>
<ul><li>
<p><code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li>accept <code>p_cutoff</code> and deprecate <code>cutoffRowMinP</code> for future removal. <code>p_cutoff</code> is used in other package functions, as is <code>min_count</code>.</li>
</ul></li>
<li>
<p><code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code></p>
<ul><li>problem: <code>style="dotplot"</code> and <code>style="heatmap"</code> have very different visual effects, the dotplot de-emphasizes significance in favor of gene count (point size) - smaller gene count hides significance. The pale bivalent colors are very close to white, while <code>"Reds"</code> color gradient is much more distinctive.</li>
<li>Can try transforming point size, so the minimum size starts out larger? Arguments <code>point_size_min=3</code> and <code>point_size_max=10</code> may help?</li>
<li>Another more dramatic option is to create normal heatmap, then just draw points on top of the already-filled cells. Note the points would be inside the boxes, instead of connected by lines through the center of each box. Could test with <code>style="hybrid"</code> and implement both.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_13oct2022">13oct2022<a class="anchor" aria-label="anchor" href="#id_13oct2022"></a></h2>
<ul><li>
<p>S4 objects?</p>
<ul><li>
<p><code>mem</code>:</p>
<ul><li>Idea is to have one object type to handle output from <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>
</li>
<li>benefit is mostly to have default behaviors for things like <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code> but also for clarity during the workflow. The <code>mem</code> object can be clear input to other functions.</li>
</ul></li>
<li>
<p><code>cnet</code>:</p>
<ul><li>inherits <code>igraph</code> and extends assumptions, again mainly for clarity as input argument to other functions</li>
</ul></li>
</ul></li>
<li>
<p><code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code></p>
<ul><li>option to display dot plot format at the top of heatmaps, equivalent to calling <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code>. Benefit would be to indicate direction (color) and number of genes (size) in addition to P-value (intensity).</li>
</ul></li>
<li>
<p>fix <code><a href="reference/reorderIgraphNodes.html">reorderIgraphNodes()</a></code></p>
<ul><li><p>DONE: add “frame.color”: <code>sortAttributes=c("pie.color", "pie.color.length", "coloredrect.color", "color", "pie.border", "frame.color", "label", "name")</code></p></li>
<li>
<p>future: make sort more intelligent, so it uses appropriate color based upon node shape during the sort.</p>
<ul><li>
<code>shape="pie"</code>: use “pie.color”, “pie.border”, “frame.color”, “label”, “name”</li>
<li>
<code>shape="coloredrect"</code>: use “coloredrect.color”, “coloredrect.border”, “frame.color”, “label”, “name”</li>
<li>all others assumed to be <code>shape="circle"</code> or similar: use “color”, “frame.color” and ignore “pie.color”, “pie.border”, “coloredrect.color”, “coloredrect.border”, “label”, “name”</li>
</ul></li>
</ul></li>
<li>
<p>DONE: Include gene direction in the workflow:</p>
<ul><li>
<p>DONE: Add <code>geneIMdirection</code> to mem object.</p>
<ul><li>Requires <code>geneHitIM</code> input to <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>.</li>
<li>Alternate backwards compatibility is function to add <code>geneIMdirection</code> to <code>mem</code> object, and/or <code>cnet</code> <code>igraph</code> object.</li>
<li>DONE: <code><a href="reference/mem2cnet.html">memIM2cnet()</a></code> should optionally use <code>enrichIMdirection</code> to apply border color.</li>
</ul></li>
<li><p>DONE: <code><a href="reference/mem2cnet.html">mem2cnet()</a></code> as its own function, minor variation and extension to <code><a href="reference/mem2cnet.html">memIM2cnet()</a></code> which only uses the <code>memIM</code> portion of the data.</p></li>
<li><p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code> should recognize <code>lwd</code> when plotting nodes. This change diverges from default behavior in <code>igraph</code> which only uses <code>par("lwd")</code> as a global adjustment for all lines while plotting nodes, therefore does not allow any nodes to have different line widths.</p></li>
<li><p>DONE: Outline of Cnet gene nodes by direction.</p></li>
<li><p>Gene-Path heatmap rows should optionally indicate direction, possibly another stripe using geneIMdirection.</p></li>
</ul></li>
<li>
<p><code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></p>
<ul><li>should call <code><a href="reference/mem2cnet.html">mem2cnet()</a></code> - or just avoid this step altogether since the <code>cnet</code> process is better done with <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>
</li>
<li>consider removing the <code>enrichMap</code> and <code>cnet</code> steps altogether.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_26sep2022">26sep2022<a class="anchor" aria-label="anchor" href="#id_26sep2022"></a></h2>
<ul><li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></p>
<ul><li>FIXED: <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> does not honor <code>do_plot=FALSE</code>.</li>
</ul></li>
<li>
<p><code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code></p>
<ul><li>DONE: add argument <code>cluster_rows</code> to allow <code>cluster_rows=FALSE</code>.</li>
<li>DONE for manual plot calls: one goal is to draw this heatmap using the order from the gene-pathway heatmap clustering.</li>
<li>Future idea: allow plotting this data using the same order as the gene-pathway heatmap. This process would require running the gene-pathway clustering first, determining the column order, then using it to order the rows in this heatmap.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_31aug2022">31aug2022<a class="anchor" aria-label="anchor" href="#id_31aug2022"></a></h2>
<ul><li>
<p>Now that <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> and node shapes <code>"jampie"</code> render the <code>pie.border</code> and <code>frame.color</code>, which can indicate the direction of change for each gene, in context of each enrichment test, some other changes should be made to the workflow:</p>
<ul><li><p>Consider adding <code>pie.lwd</code> to recognized attributes in <code><a href="reference/shape.jampie.plot.html">shape.jampie.plot()</a></code>, but confirm that this argument can be used when rendering <code>"jampie"</code> node shapes. Currently the line width can only be adjusted with <code>par("lwd")</code> which is a global setting.</p></li>
<li><p>The default <code>frame.color</code> (<code>vertex.frame.color</code>) should probably be set to <code>NA</code> or <code>"transparent"</code> so the frame color is not visible for pie nodes. It shows up as a small white line now, which was not visible previously.</p></li>
<li><p><code><a href="reference/reorderIgraphNodes.html">reorderIgraphNodes()</a></code> should include <code>pie.border</code> and <code>frame.color</code> in sensible default locations, so nodes will also be sorted by these values when relevant, without the user having to add these columns.</p></li>
<li>
<p>New function to populate <code>frame.color</code> and <code>pie.border</code> for Cnet plots.</p>
<ul><li>When <code>pie</code> has only one value, apply color to <code>frame.color</code>
</li>
<li>When <code>pie</code> has multiple values, and different directions, apply colors to <code>pie.border</code>
</li>
<li>When <code>pie</code> has multiple values, and all have the same direction, apply colors to <code>frame.color</code>
</li>
<li>In absence of any directional data, set all <code>frame.color</code> to default, and set all <code>pie.border</code> to <code>"transparent"</code> or <code>NA</code>.</li>
</ul></li>
</ul></li>
<li>
<p>Need to include gene direction of change in the workflow:</p>
<ul><li>Some easy method to include direction of change in the <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> workflow, for example argument <code>geneHitList</code> currently uses a <code>character</code> vector, but could accept <code>integer</code> vector direction with genes stored as <code>character</code> labels, same as used with <code>venndir</code> signed input lists.</li>
<li>When gene direction is available, the <code>frame.color</code> and/or <code>pie.border</code> colors are defined.</li>
<li>
<code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code> option to represent direction of change in the <code>gene_im</code> incidence matrix.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="aug2022">aug2022<a class="anchor" aria-label="anchor" href="#aug2022"></a></h2>
<ul><li>
<p>Edge bundling makes assumptions for bipartite graphs (cnet) that are difficult to use with normal graphs.</p>
<ul><li>
<code><a href="reference/edge_bundle_nodegroups.html">edge_bundle_nodegroups()</a></code> probably needs to subset edges involved in each nodegroup before bundling all edges connected to these nodes. Currently, nodes are assumed to share all connections, but the effect should occur for edges where both ends of the edge are contained in the nodegroup entries.</li>
<li>consider node attribute “<code>nodegroup</code>”.</li>
<li>consider edge attribute <code>"edgegroup"</code>, which would probably be the optimal approach for edge bundling, apart from implementing another technique similar to force-directed, hierarchically-defined, or density-directed edge bundling.</li>
<li>consider an option to specify the “midpoint” for a nodegroup, to allow some control on the spline curvature. Bonus points for allowing multiple points in order, to influence a path.</li>
</ul></li>
<li>
<p><code><a href="reference/jam_igraph.html">jam_igraph()</a></code></p>
<ul><li>fancy effect: allow edge colors to have multiple values, then interpolate color along each edge. For bundled edges, make a gradient equal to the number of line segments. For straight edges, break into <code>detail</code> number of pieces. The <code>detail</code> argument is also used by the <code>edge_bundle_*()</code> functions.</li>
</ul></li>
<li><p>subset <code>igraph</code> object by nodes, with added benefit that the <code>graph_attr(g, "layout")</code> will also be subset, if present. It is odd that the default <code><a href="https://r.igraph.org/reference/subgraph.html" class="external-link">igraph::subgraph()</a></code> function would subset the graph, leaving the layout which inevitably causes an error.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_13jul2022">13jul2022<a class="anchor" aria-label="anchor" href="#id_13jul2022"></a></h2>
<ul><li>
<p><code>mem_gene_pathway_heatmap()</code></p>
<ul><li>would be useful to have an option to subset enrichments, this option would likely be useful to <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>, and <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code>, so the effect could cascade to subsequent functions.</li>
</ul></li>
<li>
<p>new function idea? <code>subset_mem()</code></p>
<ul><li><p>Could be useful to accomplish the item above, to subset by enrichments prior to <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> related functions.</p></li>
<li><p>The challenge is that subsetting only <code>enrichIM</code> does not also update corresponding <code>memIM</code> data. Data to be updated:</p></li>
<li><p><code>geneIM</code>, <code>geneIMcolors</code>, <code>geneIMdirection</code> - simple subset by colnames</p></li>
<li><p><code>enrichIM</code>, <code>enrichIMcolors</code>, <code>enrichIMgeneCount</code> - simple subset by colnames</p></li>
<li><p><code>memIM</code> - re-create after updating <code>geneIM</code></p></li>
<li><p><code>enrichList</code> - simple subset by name</p></li>
<li>
<p>Others not necessary for <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>:</p>
<ul><li>
<code>multiEnrichMap</code>, <code>multiCnetPlot</code> should be re-created using methods in <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>. To be fair, these objects are not that useful anymore, since <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> is generally preferred.</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_05jul2022">05jul2022<a class="anchor" aria-label="anchor" href="#id_05jul2022"></a></h2>
<ul><li>
<p><code>mem_gene_pathway_heatmap()</code> when supplied with custom <code>column_split</code> throws an error when <code>cluster_columns=TRUE</code>.</p>
<ul><li>internally it uses <code><a href="https://rdrr.io/pkg/amap/man/hcluster.html" class="external-link">amap::hcluster()</a></code> to generate a dendrogram/hclust</li>
<li>using split and dendrogram together is not allowed by <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>
</li>
<li>this clustering also uses the incidence matrix combined with the pathway enrichment annotation displayed along the top, and these values are weighted with <code>pathway_column_weight</code>.</li>
<li>A proper solution would be to provide a custom function for <code>cluster_columns</code> that internally combined the incidence matrix and enrichment matrix data together prior to clustering. If this function receives a numeric matrix with proper colnames, it should work.</li>
</ul></li>
<li><p>COMPLETE: <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> argument <code>gene_row_title=NULL</code> is being passed to <code>mem_gene_pathway_heatmap()</code> and is therefore not using the default <code>row_title=letters</code>.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_23mar2022-issue-7-passing-arguments-through--in-mem_plot_folio">23mar2022 issue #7 passing arguments through … in <code>mem_plot_folio()</code>
<a class="anchor" aria-label="anchor" href="#id_23mar2022-issue-7-passing-arguments-through--in-mem_plot_folio"></a></h2>
<ul><li>User reported an error when calling <code>mem_plot_folio(mem, node_factor=5, label_factor=1.5)</code>
</li>
<li>
<code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> is passing <code>...</code> to <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code> which does not accept <code>...</code> and throws an error when receiving extra arguments.</li>
<li>I need to limit <code>...</code> to arguments accepted by <code>Heatmap()</code>. I feel like there is an R package function to handle this scenario, so I can avoid writing <code>do.call(Heatmap, custom_args)</code>.</li>
</ul></div>
<div class="section level2">
<h2 id="id_04feb2022">04feb2022<a class="anchor" aria-label="anchor" href="#id_04feb2022"></a></h2>
<p>Workflow that starts with pathway-gene incidence matrix upfront, bypassing the <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> workflow altogether.</p>
<ul><li>
<p><code><a href="reference/mem2cnet.html">memIM2cnet()</a></code> to convert pathway-to-gene incidence matrix to Cnet <code>igraph</code>.</p>
<ul><li>Requires <code>geneIM</code> which is the enrichment-to-gene incidence matrix. In the driving example, each enrichment would represent a disease subgroup, with the full set of differential genes associated to each subgroup.</li>
<li>Requires <code>enrichIM</code> which contains enrichment-to-pathway whose values are P-values. When this data is not supplied, it should use 0.001 by default. I think these values are only used as an optional gradient color for the pathway nodes.</li>
<li>Optional <code>geneIMcolors</code> which is the same as <code>geneIM</code> except the cells contain colors to use for each gene. Currently the function does not fill these colors, the best method is to use <code><a href="https://jmw86069.github.io/colorjam/reference/matrix2heatColors.html" class="external-link">colorjam::matrix2heatColors()</a></code>
</li>
</ul><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneIMcolors</span> <span class="op">&lt;-</span> <span class="fu">colorjam</span><span class="fu">::</span><span class="fu"><a href="https://jmw86069.github.io/colorjam/reference/matrix2heatColors.html" class="external-link">matrix2heatColors</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">geneIM</span>,</span>
<span>   colorV<span class="op">=</span><span class="va">colorV</span><span class="op">)</span></span></code></pre></div>
</li>
</ul></div>
<div class="section level2">
<h2 id="id_03feb2022">03feb2022<a class="anchor" aria-label="anchor" href="#id_03feb2022"></a></h2>
<ul><li>
<p>option to assign pathways to “functional themes” based upon presentation by Adeline Chin in Dr. Hanna Kim’ group.</p>
<ul><li>This step may “collapse” multiple pathways together, similar to grouping functional groups: union of genes, lowest enrichment P-value.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_25jan2021">25jan2021<a class="anchor" aria-label="anchor" href="#id_25jan2021"></a></h2>
<ul><li>
<p><code>heatmap_row_order()</code>, <code>heatmap_column_order()</code> should return a flat vector when there are no row_split or column_split, respectively. Currently data is returned as a list of one-length vectors.</p>
<ul><li>Consider moving into <code>jamba</code> package, in case this function needs to be re-used, it should not require loading the full <code>multienrichjam</code> package, which itself requires things like <code>igraph</code>, <code>clusterProfiler</code>, <code>qgraph</code>, <code>DOSE</code>, <code>matrixStats</code>.</li>
<li>Import <code><a href="https://jmw86069.github.io/jamba/reference/heatmap_row_order.html" class="external-link">jamba::heatmap_row_order()</a></code> and <code><a href="https://jmw86069.github.io/jamba/reference/heatmap_column_order.html" class="external-link">jamba::heatmap_column_order()</a></code> to ensure any functions or packages calling this function will succeed without error.</li>
</ul></li>
</ul><p>This document describes plans for enhancements to the multienrichjam R package.</p>
</div>
<div class="section level2">
<h2 id="id_05nov2021">05nov2021<a class="anchor" aria-label="anchor" href="#id_05nov2021"></a></h2>
<p>Now that directional z-score can also be associated with enrichment P-values, heatmaps might need to use a bivariate color scale, to indicate enrichment and directionality. See “stevens.bluered”. For example the <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> colors nodes by enrichment P-value, more intense is more significant enrichment. The z-score direction is used to apply red “activated” or blue “inhibited”. However, pathways with no z-score, or z-score below the threshold are colored red by default. They should use a neutral color.</p>
</div>
<div class="section level2">
<h2 id="id_04nov2021">04nov2021<a class="anchor" aria-label="anchor" href="#id_04nov2021"></a></h2>
<p>For IPA “Upstream Regulators” it sometimes offers a direction implied by the <code>activation z-score</code>. Design idea is to implement the directionality so it can be included in downstream analyses.</p>
<ul><li>
<code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> - currently shades by the <code>-log10(pvalue)</code> however if there is directionality, it could be signed <code>+</code> for activated, <code>-</code> for inhibited. Then the heatmap color scale would use blue-white-red color gradient.</li>
<li>
<code>mem$enrichIMdirection</code> contains matrix of direction, by default <code>1</code> means all have same direction.</li>
</ul><p>During <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> it filters for <code>topEnrichN</code> entries for each enrichment. It might be useful to retain the rank number for each enrichment, to review when setting a different <code>topEnrichN</code> threshold.</p>
<ul><li>
<code>mem$enrichIMgeneCount</code> contains matrix of gene counts</li>
<li>
<code>mem$enrichIMrank</code> contains matrix of pathway rank (after filtering gene count)</li>
</ul></div>
<div class="section level2">
<h2 id="id_01nov2021">01nov2021<a class="anchor" aria-label="anchor" href="#id_01nov2021"></a></h2>
<ul><li>
<p>COMPLETE: <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> - the heatmap circles and legend circles are not the same size - they should be fixed to the same absolute size.</p>
<ul><li>COMPLETE: Optionally label each heatmap cell with the number of genes for visual reinforcement.</li>
</ul></li>
<li>
<p>COMPLETE: When there are more than 3 enrichments, the color legend on the gene-pathway heatmap becomes unwieldy - taking over the whole figure.</p>
<ul><li>Optionally (and by default) hide color legend for the gene-pathway heatmap.</li>
</ul></li>
<li>
<p>COMPLETE: The gene-pathway heatmap use_raster=TRUE causes artifacts in output, it should be disabled by default. In future, debug why things go wrong.</p>
<ul><li>I think the bug is caused by rasterization being done on underlying numbers before the color ramp is applied, in which case this problem cannot be solved when <code>colorize_by_gene=TRUE</code> since categorical values are assigned numbers which are not a proper color gradient.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_20jul2021">20jul2021<a class="anchor" aria-label="anchor" href="#id_20jul2021"></a></h2>
<ul><li>
<p>Replace <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> with <code><a href="reference/multienrichjam-package.html">multienrichjam()</a></code> and simplify the arguments:</p>
<ul><li>p_cutoff</li>
<li>min_gene_count</li>
<li>top_enrich_n</li>
<li>colnames: id, name, description, pvalue, gene</li>
<li>color_sub</li>
</ul></li>
<li><p>Port <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> argument <code>colorize_by_gene=TRUE</code> to use <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code> instead of <code><a href="https://jmw86069.github.io/jamba/reference/imageByColors.html" class="external-link">jamba::imageByColors()</a></code> for consistency, also so it can support <code>style="dotplot"</code>.</p></li>
</ul></div>
<div class="section level2">
<h2 id="id_19jul2021">19jul2021<a class="anchor" aria-label="anchor" href="#id_19jul2021"></a></h2>
<ul><li>
<code><a href="reference/jam_igraph.html">jam_igraph()</a></code> with <code>rescale=TRUE</code> should also scale igraph vertex size and igraph label size according to the new axis ranges.</li>
<li>Debug <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> when <code>vertex.size</code> is defined alongside V(g)$size, and <code>node_factor</code>. It appears not to apply the size properly.</li>
<li>COMPLETE: <code><a href="reference/mem_enrichment_heatmap.html">mem_enrichment_heatmap()</a></code> new option for dot plot format, based upon <code><a href="https://rdrr.io/pkg/enrichplot/man/dotplot.html" class="external-link">enrichplot::dotplot()</a></code> that sizes each dot by the number of genes present.</li>
</ul></div>
<div class="section level2">
<h2 id="id_24jun2021">24jun2021<a class="anchor" aria-label="anchor" href="#id_24jun2021"></a></h2>
<ul><li>In <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> remove default <code>topEnrichSources</code> and <code>topEnrichSourceSubset</code> which throw errors when not using MSigDB data.</li>
<li>
<code><a href="reference/topEnrichBySource.html">topEnrichBySource()</a></code> and <code><a href="reference/topEnrichBySource.html">topEnrichListBySource()</a></code> should be able to accept <code>enrichResult</code> as input, and return <code>enrichResult</code> and not <code>data.frame</code> which is understandably lossy. I need to understand the proper method for creating a subset of an <code>enrichList</code> object, including its internal data.</li>
<li>Streamline the <code><a href="reference/topEnrichBySource.html">topEnrichListBySource()</a></code> workflow.</li>
</ul></div>
<div class="section level2">
<h2 id="id_21jun2021">21jun2021<a class="anchor" aria-label="anchor" href="#id_21jun2021"></a></h2>
<ul><li>COMPLETE: Allow rotating gene-pathway incidence matrix when using <code><a href="reference/mem_gene_path_heatmap.html">mem_gene_path_heatmap()</a></code>, so the pathway names are rows.</li>
<li>Add test data object <code>mem</code> to be convenient for function examples, and testthis test suite.</li>
</ul></div>
<div class="section level2">
<h2 id="smaller-usability-items">Smaller usability items<a class="anchor" aria-label="anchor" href="#smaller-usability-items"></a></h2>
<ul><li>
<p>edge_bundling=“connections” should also allow <code>render_groups=TRUE</code> to work, by returning the <code>"nodegroups"</code> required for that step.</p>
<ul><li>It should probably filter out singlet nodes by default, since the typical use case is “nodeset-to-node” for Cnet plots.</li>
<li>Main use case is with a Cnet plot, edge bundling by connected nodes should allow drawing an optional border around each group of nodes.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="complete-10apr2021-implement-edge-bundling-techniques">COMPLETE: 10apr2021: implement edge bundling techniques<a class="anchor" aria-label="anchor" href="#complete-10apr2021-implement-edge-bundling-techniques"></a></h2>
<p>Priority: high</p>
<p>Status: Implemented, early active testing of usability and functionality</p>
<ul><li>“connections” - Cnet edge bundles - special case of nodeset-to-node edge bundling</li>
<li>“nodegroups”: Node group bundles - general case of nodeset-to-nodeset edge bundling</li>
</ul><p>Useful to improve readability/aesthetics of collapsed Cnet plots, especially Cnet plots with many gene nodes where it is difficult to tell which pathway clusters are connected to each gene.</p>
</div>
<div class="section level2">
<h2 id="id_10apr2021-enhancements-to-jam_igraph-must-be-done-inside-jam_plot_igraph">10apr2021: enhancements to <code>jam_igraph()</code> (must be done inside <code>jam_plot_igraph()</code>)<a class="anchor" aria-label="anchor" href="#id_10apr2021-enhancements-to-jam_igraph-must-be-done-inside-jam_plot_igraph"></a></h2>
<ul><li><p>Consider silently returning the <code>igraph</code> object plotted where the vertex, edge, and graph attributes are updated to the values used at runtime. For example, if user overrides any attributes, those attributes will be present in the object returned, so the next iteration someone could just call <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> without any custom attributes and it would produce the same plot again.</p></li>
<li><p>See <a href="https://igraph.org/r/doc/plot.common.html" class="external-link uri">https://igraph.org/r/doc/plot.common.html</a> and <code><a href="https://r.igraph.org/reference/igraph_options.html" class="external-link">igraph::igraph_options()</a></code></p></li>
</ul></div>
<div class="section level2">
<h2 id="id_10apr2021-implement-edge-bundling-in-jam_igraph-and-jam_plot_igraph">10apr2021: implement edge bundling in <code>jam_igraph()</code> and <code>jam_plot_igraph()</code>
<a class="anchor" aria-label="anchor" href="#id_10apr2021-implement-edge-bundling-in-jam_igraph-and-jam_plot_igraph"></a></h2>
<p>Priority: high</p>
<p>Status: Implemented</p>
<ul><li>
<p>This task involves making edge bundling accessible as a simple option during plotting, to prevent having to run 3 or 4 functions:</p>
<ul><li>
<code><a href="reference/jam_igraph.html">jam_igraph()</a></code> with hidden edges</li>
<li>
<code>bundle_node_edges()</code> to display edges on top of nodes</li>
<li>(optional) <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> with hidden edges, nodes with solid white color so they fully cover edges.</li>
<li>
<code><a href="reference/jam_igraph.html">jam_igraph()</a></code> with hidden edges, to display nodes on top of bundled edges while allowing nodes to have alpha transparency</li>
</ul></li>
<li>
<p>Note there is a package based on <code>ggraph</code> that implements edge bundling - if <code>jam_ggraph()</code> is going to be our future, maybe we should implement edge bundling using a similar approach used by that R package.</p>
<ul><li><a href="https://github.com/schochastics/edgebundle" class="external-link uri">https://github.com/schochastics/edgebundle</a></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10apr2021-add-plot_cnet_heatmaps-to-multienrichjam">10apr2021: add <code>plot_cnet_heatmaps()</code> to multienrichjam<a class="anchor" aria-label="anchor" href="#id_10apr2021-add-plot_cnet_heatmaps-to-multienrichjam"></a></h2>
<p>Priority: high</p>
<ul><li>
<code>plot_cnet_heatmaps()</code> is in development, and arranges expression heatmaps around a central Cnet plot, using genes in each cnet cluster. It is intended to be used with Cnet clusters from <code><a href="reference/collapse_mem_clusters.html">collapse_mem_clusters()</a></code>, by-product of <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code>.</li>
</ul><p>This figure seems very useful because it integrates expression changes alongside gene-pathway connections.</p>
<p>The multienrichment workflow would likely become:</p>
<ul><li>Prepare enrichment data</li>
<li><code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code></li>
<li><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code></li>
<li><code>plot_cnet_heatmap()</code></li>
</ul></div>
<div class="section level2">
<h2 id="id_10apr2021-enable-plotting-node-community-data-in-jam_igraph">10apr2021: enable plotting node community data in <code>jam_igraph()</code>
<a class="anchor" aria-label="anchor" href="#id_10apr2021-enable-plotting-node-community-data-in-jam_igraph"></a></h2>
<p>Priority: low</p>
<ul><li>Analogous to <code>igraph:::plot.communities()</code> with <code>mark.groups</code>
</li>
<li>for edge bundling, this step could optionally display boundaries around each node group.</li>
</ul><p>Value is reinforcing node groups with a boundary. In testing, the boundary actually made plots look more complex. And with simple plots, it does not seem necessary since nodes are already well-spaced.</p>
</div>
<div class="section level2">
<h2 id="id_10apr2021-include-genecount-matrix-output-from-multienrichmap">10apr2021: include geneCount matrix output from <code>multiEnrichMap()</code>
<a class="anchor" aria-label="anchor" href="#id_10apr2021-include-genecount-matrix-output-from-multienrichmap"></a></h2>
<p>Priority: medium</p>
<ul><li><p>Purpose is to complement the “enrichIM” matrix, with P-values for each set. However “dysregulated pathways” may also be required to have N number of genes, yet this data is not easily available.</p></li>
<li>
<p>Rows are pathways, columns are enrichments, values are the number of genes involved in enrichment.</p>
<ul><li>Consider a matrix whose cells contain delimited gene symbols</li>
</ul></li>
</ul><p>Useful to apply filtering at matrix-level for pathways that meet enrichment P-value and gene count thresholds.</p>
</div>
<div class="section level2">
<h2 id="id_2020-apply-node-color-order-based-upon-shape-of-node-cluster">2020: apply node color order based upon shape of node cluster<a class="anchor" aria-label="anchor" href="#id_2020-apply-node-color-order-based-upon-shape-of-node-cluster"></a></h2>
<p>Priority: medium</p>
<p><code><a href="reference/reorderIgraphNodes.html">reorderIgraphNodes()</a></code></p>
<ul><li><p>when a cluster of nodes is short-wide, the order should be left-to-right</p></li>
<li><p>when a cluster of nodes is tall-skinny, the order should be top-to-bottom</p></li>
<li>
<p>all else should be sorted left-to-right (or user-defined default order)</p>
<ul><li>it is visually confusing when tall-skinny nodes are sorted left-to-right</li>
</ul></li>
</ul><p>Useful to automate the ordering of node colors</p>
</div>
<div class="section level2">
<h2 id="id_2020-apply-subsetsets-in-multienrichmap">2020: apply subsetSets in <code>multiEnrichMap()</code>
<a class="anchor" aria-label="anchor" href="#id_2020-apply-subsetsets-in-multienrichmap"></a></h2>
<p>Priority: low</p>
<ul><li>apparently it had never been implemented? User-defined subset of pathways/sets to include in <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>, as an alternative to using <code><a href="reference/topEnrichBySource.html">topEnrichBySource()</a></code>.</li>
</ul><p>Users can currently perform this subset step before running <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="complete-apply-min_count-in-multienrichmap">COMPLETE: Apply min_count in <code>multiEnrichMap()</code>
<a class="anchor" aria-label="anchor" href="#complete-apply-min_count-in-multienrichmap"></a></h2>
<ul><li>Currently <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code> does not filter by number of genes involved in enrichment, it only filters by enrichment P-value. New argument <code>min_count</code> is applied only when <code>topEnrichN</code> is used, but nothing else downstream is aware of filtering by <code>min_count</code>. The corresponding argument in <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> is <code>min_set_ct_each</code>, which requires a set (pathway) to contain at least this many entries in at least one enrichment result which also meets <code>p_cutoff</code> criteria for enrichment P-value.</li>
</ul></div>
<div class="section level2">
<h2 id="id_2020-optional-highlight-genes">2020: Optional highlight genes<a class="anchor" aria-label="anchor" href="#id_2020-optional-highlight-genes"></a></h2>
<p>Priority: low</p>
<ul><li>
<p><code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> and subsequent plots, optional argument <code>highlight_genes</code> which would effectively hide all gene labels except <code>highlight_genes</code> – to help especially crowded plots.</p>
<ul><li>Option 1: User-supplied genes</li>
<li>Option 2: Genes defined by one or more pathways/sets to highlight</li>
<li>Implement as wrapper to <code><a href="reference/jam_igraph.html">jam_igraph()</a></code> to hide/display relevant node labels</li>
<li>
<code>plot_cnet_heatmap()</code> could label the heatmap rows using <code>anno_mark()</code>
</li>
</ul></li>
</ul><p>Useful in Cnet plots with too many genes to label, it allows only a subset of genes to be highlighted. Will become high priority if a manuscript decides to use this technique.</p>
</div>
<div class="section level2">
<h2 id="consider-a-strict-output-format-for-mem_plot_folio">Consider a strict output format for <code>mem_plot_folio()</code>
<a class="anchor" aria-label="anchor" href="#consider-a-strict-output-format-for-mem_plot_folio"></a></h2>
<p>Priority: medium</p>
<ul><li><p>Goal is to run <code><a href="reference/mem_plot_folio.html">mem_plot_folio()</a></code> once then be able to plot any component separately. Would help keep all plots in sync with the settings used.</p></li>
<li><p>Benefit: Always have ability to see the exact gene-pathway incidence matrix heatmap, and its clustering, which is used for a collapsed Cnet plots.</p></li>
<li>
<p>Design idea: List format</p>
<ul><li>
<p>Named by type of output:</p>
<ul><li><p>enrichment P-values</p></li>
<li>
<p>gene-pathway heatmap</p>
<ul><li>Heatmap object</li>
<li>gene clusters</li>
<li>pathway clusters</li>
<li>filter/cluster settings used</li>
</ul></li>
<li>
<p>collapsed cnet igraphs</p>
<ul><li>with each labeling option implemented</li>
</ul></li>
<li>
<p>cnet exemplar igraphs</p>
<ul><li>1 exemplar per pathway cluster</li>
<li>2 exemplars per pathway cluster</li>
<li>3 exemplars per pathway cluster</li>
</ul></li>
<li>
<p>each cnet cluster igraph</p>
<ul><li>cluster 1 cnet igraph</li>
<li>cluster 2 cnet igraph</li>
<li>cluster 3 cnet igraph</li>
<li>…</li>
</ul></li>
</ul></li>
<li><p>Each type of output contains a <code>list</code> of relevant components</p></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10apr2021-make-plot-commands-word-for-ggraph">10apr2021: make plot commands word for ggraph<a class="anchor" aria-label="anchor" href="#id_10apr2021-make-plot-commands-word-for-ggraph"></a></h2>
<p>Priority: low</p>
<ul><li><p>Goal is to use ggraph for ggplot2-type plotting instead of using base R for <code>igraph</code> plots.</p></li>
<li>
<p><code>jam_ggraph()</code> ?</p>
<ul><li>Test <code>scatterpie</code> R package which implements pie node shapes for use in ggraph. Unsure the data content requirement.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="id_10apr2021-document-examples-for-reorderigraphnotes">10apr2021: document examples for <code>reorderIgraphNotes()</code>
<a class="anchor" aria-label="anchor" href="#id_10apr2021-document-examples-for-reorderigraphnotes"></a></h2>
<ul><li>Examples should show how to specify color order, or change the order if necessary.</li>
</ul></div>
<div class="section level2">
<h2 id="complete-2020-improve-reorderigraphnotes">COMPLETE: 2020: Improve <code>reorderIgraphNotes()</code>
<a class="anchor" aria-label="anchor" href="#complete-2020-improve-reorderigraphnotes"></a></h2>
<ul><li>
<code><a href="reference/reorderIgraphNodes.html">reorderIgraphNodes()</a></code> when it encounters attributes with multiple colors per node, such as <code>"pie.colors"</code> and <code>"coloredrect.colors"</code>, calls <code><a href="reference/avg_colors_by_list.html">avg_colors_by_list()</a></code> to generate one blended color per node, then it sorts those colors using <code>sort_colors()</code>. However, the color order ultimately does not match the order in the color legend, and other plots such as heatmaps.</li>
<li>Future idea is to convert node attributes with multiple colors per node into a <code>data.frame</code> where each color is in a separate column. Then convert each column to a factor whose levels match the color order from <code>colorV</code> (argument to <code><a href="reference/multiEnrichMap.html">multiEnrichMap()</a></code>). The end result should sort nodes consistent with the order of colors.</li>
<li>Note this process assumes that all node colors are from a limited set, which ideally should match <code>colorV</code>. Therefore if any color gradient is applied to nodes with <code>nodeType="gene"</code>, this process will not work.</li>
</ul></div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

