<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Multienrichment with clusterProfiler • multienrichjam</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Multienrichment with clusterProfiler">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">multienrichjam</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.103.900</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/importClusterProfiler.html">Multienrichment with clusterProfiler</a></li>
    <li><a class="dropdown-item" href="../articles/importIPA.html">How to use Ingenuity IPA enrichment results</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Explanation</h6></li>
    <li><a class="dropdown-item" href="../articles/MemPlotConcepts.html">Mem Plot Concepts</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="http://github.com/jmw86069/multienrichjam/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="importClusterProfiler_files/kePrint-0.0.1/kePrint.js"></script><link href="importClusterProfiler_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Multienrichment with clusterProfiler</h1>
            
      
      <small class="dont-index">Source: <a href="http://github.com/jmw86069/multienrichjam/blob/HEAD/vignettes/importClusterProfiler.Rmd" class="external-link"><code>vignettes/importClusterProfiler.Rmd</code></a></small>
      <div class="d-none name"><code>importClusterProfiler.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jmw86069.github.io/multienrichjam/">multienrichjam</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jmw86069.github.io/jamba/" class="external-link">jamba</a></span><span class="op">)</span>;</span>
<span><span class="co"># library(colorjam);</span></span>
<span><span class="co"># suppressPackageStartupMessages(library(ComplexHeatmap))</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"warn"</span><span class="op">=</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="msigdbr-requirement">msigdbr Requirement<a class="anchor" aria-label="anchor" href="#msigdbr-requirement"></a>
</h2>
<p>This guide requires the <code>msigdbr</code> R package from CRAN.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"msigdbr"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>   <span class="fu">jamba</span><span class="fu">::</span><span class="fu"><a href="https://jmw86069.github.io/jamba/reference/printDebug.html" class="external-link">printDebugHtml</a></span><span class="op">(</span><span class="st">"The "</span>, <span class="st">"msigdbr"</span>,</span>
<span>      <span class="st">" package is required for this vignette. Stopping here."</span><span class="op">)</span></span>
<span>   <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/knit_exit.html" class="external-link">knit_exit</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="clusterprofiler-enrichment">clusterProfiler enrichment<a class="anchor" aria-label="anchor" href="#clusterprofiler-enrichment"></a>
</h2>
<p>This document describes steps recommended for clusterProfiler
enrichment data, which include specific objects such as
<code>enrichResults</code> and others.</p>
<p>Refer to the <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/index.html" class="external-link">clusterProfiler
e-Book</a>, an outstanding and comprehensive guide to using
clusterProfiler to generate gene set enrichment data.</p>
<p>There are two general approaches:</p>
<ol style="list-style-type: decimal">
<li>Run <code><a href="https://rdrr.io/pkg/clusterProfiler/man/enricher.html" class="external-link">clusterProfiler::enricher()</a></code> for a <code>list</code>
of experiments.</li>
<li>Use an existing <code>list</code> of
<code><a href="https://rdrr.io/pkg/clusterProfiler/man/enricher.html" class="external-link">clusterProfiler::enricher()</a></code> results.</li>
</ol>
</div>
<div class="section level2">
<h2 id="clusterprofiler-enrichment-1">clusterProfiler enrichment<a class="anchor" aria-label="anchor" href="#clusterprofiler-enrichment-1"></a>
</h2>
<div class="section level3">
<h3 id="prepare-msigdb-pathway-data">Prepare MSigDB Pathway Data<a class="anchor" aria-label="anchor" href="#prepare-msigdb-pathway-data"></a>
</h3>
<p>The example below demonstrates how to prepare canonical pathways from
<a href="https://www.gsea-msigdb.org/gsea/msigdb" class="external-link">MSigDB</a> to use for
gene set enrichment. These pathways are used with a list of genes in
<code>Reese_genes</code> to test for enrichment.</p>
<p>The <code>msigdbr</code> R package is used to download necessary
data, following the guidance in the clusterProfiler e-Book: <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html#msigdb-analysis" class="external-link">MSigDb
analysis</a>. See <code><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr::msigdbr()</a></code> for more details.</p>
<div class="section level4">
<h4 id="review-msigdb-collections">Review MSigDB Collections<a class="anchor" aria-label="anchor" href="#review-msigdb-collections"></a>
</h4>
<p>The <code>msigdbr</code> package offers convenient access to
collections of gene sets available from MSigDB, shown below.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">msigdb_collections</span> <span class="op">&lt;-</span> <span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr_collections.html" class="external-link">msigdbr_collections</a></span><span class="op">(</span>db_species<span class="op">=</span><span class="st">"HS"</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<caption>
MSigDB Collections
</caption>
<thead><tr>
<th style="text-align:left;">
gs_collection
</th>
<th style="text-align:left;">
gs_subcollection
</th>
<th style="text-align:left;">
gs_collection_name
</th>
<th style="text-align:right;">
num_genesets
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C1
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Positional
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
302
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CGP
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Chemical and Genetic Perturbations
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
3,538
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CP
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Canonical Pathways
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
19
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CP:BIOCARTA
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BioCarta Pathways
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
292
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CP:KEGG_LEGACY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
KEGG Legacy Pathways
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
186
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CP:KEGG_MEDICUS
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
KEGG Medicus Pathways
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
658
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CP:PID
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
PID Pathways
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
196
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CP:REACTOME
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Reactome Pathways
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
1,787
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C2
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CP:WIKIPATHWAYS
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
WikiPathways
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
885
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C3
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
MIR:MIRDB
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
miRDB
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
2,377
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C3
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
MIR:MIR_LEGACY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
MIR_Legacy
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
221
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C3
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
TFT:GTRD
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
GTRD
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
505
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C3
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
TFT:TFT_LEGACY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
TFT_Legacy
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
610
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C4
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
3CA
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Curated Cancer Cell Atlas gene sets
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
148
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C4
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CGN
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Cancer Gene Neighborhoods
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
427
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C4
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CM
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Cancer Modules
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
431
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C5
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
<a href="GO:BP" class="uri">GO:BP</a>
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
GO Biological Process
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
7,583
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C5
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
<a href="GO:CC" class="uri">GO:CC</a>
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
GO Cellular Component
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
1,042
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C5
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
<a href="GO:MF" class="uri">GO:MF</a>
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
GO Molecular Function
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
1,855
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C5
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
HPO
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Human Phenotype Ontology
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
5,748
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C6
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Oncogenic Signature
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
189
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C7
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
IMMUNESIGDB
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
ImmuneSigDB
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
4,872
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C7
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
VAX
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
HIPC Vaccine Response
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
347
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
C8
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Cell Type Signature
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
866
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
H
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
Hallmark
</td>
<td style="text-align:right;border-left:1px solid #DDDDDD;white-space: nowrap;">
50
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="msigdb-canonical-pathways">MSigDB Canonical Pathways<a class="anchor" aria-label="anchor" href="#msigdb-canonical-pathways"></a>
</h4>
<p>This tutorial uses <code>collection = "C2"</code> because it contains
all canonical pathway gene sets from multiple sources. The canonical
pathways are then filtered by retaining the subset with
<code>gs_subcollection</code> containing <code>"CP"</code>.</p>
<p>There are two columns used by
<code><a href="https://rdrr.io/pkg/clusterProfiler/man/enricher.html" class="external-link">clusterProfiler::enricher()</a></code>:</p>
<ol style="list-style-type: decimal">
<li>gs_name - the gene set name</li>
<li>gene - usually gene symbol</li>
</ol>
<p>Using these two columns, the <code>data.frame</code> needs to retain
only unique rows.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># C2 canonical pathways</span></span>
<span><span class="va">msig_cp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span></span>
<span>   <span class="fu">msigdbr</span><span class="fu">::</span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span></span>
<span>      species <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>      collection<span class="op">=</span><span class="st">"C2"</span><span class="op">)</span>,</span>
<span>   <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"CP"</span>, <span class="va">gs_subcollection</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">msig_cp_gs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">msig_cp</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gs_name"</span>, <span class="st">"gene_symbol"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">msig_cp_gs</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p>For the purpose of this vignette, a subset of canonical pathways are
available using <code>data(msig_test)</code>, and should only be used
for this analysis.</p>
<table class="table table" style="margin-left: auto; margin-right: auto;">
<caption>
MSigDB Canonical Pathways
</caption>
<thead><tr>
<th style="text-align:left;">
gs_name
</th>
<th style="text-align:left;">
gene_symbol
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_AGPCR_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
ARRB1
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_AKAP13_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
AKAP13
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_AKAP95_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
AKAP8
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_AKAPCENTROSOME_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
AKAP9
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_BAD_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
ADCY1
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_CARM1_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
CARM1
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_CASPASE_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
APAF1
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_CELL2CELL_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
ACTN1
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_CERAMIDE_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
AIFM1
</td>
</tr>
<tr>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
BIOCARTA_CFTR_PATHWAY
</td>
<td style="text-align:left;border-left:1px solid #DDDDDD;white-space: nowrap;">
ADCY1
</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level3">
<h3 id="run-enricher">Run enricher()<a class="anchor" aria-label="anchor" href="#run-enricher"></a>
</h3>
<p>The <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/universal-api.html#msigdb-ora" class="external-link">clusterProfiler
documentation</a> for <code>enricher()</code> is straightforward for
over-representation analysis (ORA). Note that other clusterProfiler
enrich* tools can be used, for example
<code><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichKEGG.html" class="external-link">clusterProfiler::enrichKEGG()</a></code>,
<code><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichPC.html" class="external-link">clusterProfiler::enrichPC()</a></code>.</p>
<blockquote>
<p>The most important argument to include:</p>
<p><code>pvalueCutoff=1</code></p>
<p>This option retains all enrichment results without filtering.<br>
The P-value will be filtered later by multienrichjam.</p>
</blockquote>
<p>This example uses <code>Reese_genes</code> containing genes
identified by <a href="https://doi.org/10.1016/j.jaci.2018.11.043" class="external-link">Reese
<em>et al</em> 2019</a> in <strong>Epigenome-wide meta-analysis of DNA
methylation and childhood asthma</strong> <a href="https://doi.org/10.1016/j.jaci.2018.11.043" class="external-link uri">https://doi.org/10.1016/j.jaci.2018.11.043</a>.</p>
<p>The data are stored as a <code>list</code> of significant genes, so
we iterate the list using <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Gene hit lists</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Reese_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># enricher() for each element of a list</span></span>
<span><span class="va">erlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">Reese_genes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">igenes</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="va">er</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enricher.html" class="external-link">enricher</a></span><span class="op">(</span><span class="va">igenes</span>,</span>
<span>      pvalueCutoff<span class="op">=</span><span class="fl">1</span>,</span>
<span>      TERM2GENE<span class="op">=</span><span class="va">msig_test</span>,</span>
<span>      minGSSize<span class="op">=</span><span class="fl">5</span>, maxGSSize<span class="op">=</span><span class="fl">5000</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>You may also run a specific enrichment function in
<code>clusterProfiler</code> such as
<code><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichPC.html" class="external-link">clusterProfiler::enrichPC()</a></code> which automatically uses
<strong>Pathway Commons</strong> pathways.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Reese_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Optionally run enrichPC() which tests PathwayCommons</span></span>
<span><span class="va">erlist2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">Reese_genes</span>, <span class="kw">function</span><span class="op">(</span><span class="va">igenes</span><span class="op">)</span><span class="op">{</span></span>
<span>   <span class="va">er</span> <span class="op">&lt;-</span> <span class="fu">clusterProfiler</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/clusterProfiler/man/enrichPC.html" class="external-link">enrichPC</a></span><span class="op">(</span><span class="va">igenes</span>,</span>
<span>      pvalueCutoff<span class="op">=</span><span class="fl">1</span>,</span>
<span>      minGSSize<span class="op">=</span><span class="fl">5</span>, maxGSSize<span class="op">=</span><span class="fl">5000</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="run-multienrichmap">Run multiEnrichMap()<a class="anchor" aria-label="anchor" href="#run-multienrichmap"></a>
</h2>
<p>The <code>erlist</code> from the previous step will be the input to
<code><a href="../reference/multiEnrichMap.html">multiEnrichMap()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mem</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/multiEnrichMap.html">multiEnrichMap</a></span><span class="op">(</span><span class="va">erlist</span>,</span>
<span>   pvalueColname<span class="op">=</span><span class="st">"qvalue"</span>,</span>
<span>   p_cutoff<span class="op">=</span><span class="fl">0.01</span>,</span>
<span>   cutoffRowMinP<span class="op">=</span><span class="fl">0.2</span>,</span>
<span>   min_count<span class="op">=</span><span class="fl">2</span>,</span>
<span>   topEnrichN<span class="op">=</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>The default summary for <code>mem</code> describes the contents,
shown below:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mem</span></span>
<span><span class="co">#&gt; class: Mem</span></span>
<span><span class="co">#&gt; dim: 2 enrichments, 14 sets, 17 genes</span></span>
<span><span class="co">#&gt; - enrichments (2): Newborns, OlderChildren</span></span>
<span><span class="co">#&gt; - sets (14): REACTOME_TAK1_DEPENDENT_IKK_AND_NF_KAPPA_B_ACTIVATION, KEGG_APOPTOSIS, ..., KEGG_ASTHMA, WP_HEAD_AND_NECK_SQUAMOUS_CELL_CARCINOMA</span></span>
<span><span class="co">#&gt; - genes (17): ACTN1, ALPK1, ..., RUNX1, SPP2</span></span>
<span><span class="co">#&gt; Analysis parameters:</span></span>
<span><span class="co">#&gt; - top N per enrichment: 20</span></span>
<span><span class="co">#&gt; - significance threshold: 0.2 (colname: qvalue)</span></span>
<span><span class="co">#&gt; - min gene count: 2</span></span>
<span><span class="co">#&gt; - direction colname: zScore</span></span></code></pre></div>
<div class="section level3">
<h3 id="mem-plot-folio">Mem Plot Folio<a class="anchor" aria-label="anchor" href="#mem-plot-folio"></a>
</h3>
<p>The <code><a href="../reference/mem_plot_folio.html">mem_plot_folio()</a></code> represents a key step in the
analysis workflow.</p>
<p>Pathway clusters are defined by analyst parameters:</p>
<ul>
<li>The number of pathways clusters</li>
<li>The relative weight of the gene-pathway incidence matrix.</li>
<li>The method used for clustering.</li>
</ul>
<p>Mem Plot Folio then provides a series of visualizations, described in
detail in <code><a href="../reference/mem_plot_folio.html">mem_plot_folio()</a></code>.</p>
<p><strong>Only the first four plots are shown below</strong> using
<code>do_which=c(1, 2, 3, 4)</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Mpf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mem_plot_folio.html">mem_plot_folio</a></span><span class="op">(</span><span class="va">mem</span>,</span>
<span>   pathway_column_split<span class="op">=</span><span class="fl">4</span>,</span>
<span>   column_cex<span class="op">=</span><span class="fl">0.4</span>, row_cex<span class="op">=</span><span class="fl">0.4</span>,</span>
<span>   row_names_max_width<span class="op">=</span><span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">9</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>   column_names_max_height<span class="op">=</span><span class="fu">grid</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>   node_factor<span class="op">=</span><span class="fl">2.5</span>,</span>
<span>   label_factor_l<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nodeType<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Set<span class="op">=</span><span class="fl">0.7</span>, Gene<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span>,</span>
<span>   use_shadowText<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>   do_which<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>,</span>
<span>   main<span class="op">=</span><span class="st">"Canonical Pathways"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required namespace: gridtext</span></span></code></pre></div>
<p><img src="importClusterProfiler_files/figure-html/mem-folio-1.png" alt="mem_plot_folio, plot 1, enrichment heatmap" width="100%" style="display: block; margin: auto;"><img src="importClusterProfiler_files/figure-html/mem-folio-2.png" alt="mem_plot_folio, plot 2, Gene-Pathway heatmap" width="100%" style="display: block; margin: auto;"><img src="importClusterProfiler_files/figure-html/mem-folio-3.png" alt="mem_plot_folio, plot 3, Cnet plot with Cluster Letters" width="100%" style="display: block; margin: auto;"><img src="importClusterProfiler_files/figure-html/mem-folio-4.png" alt="mem_plot_folio, plot 4, Cnet plot with Cluster Summary Labels" width="100%" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="cnet-collapsed-cluster-plot">Cnet Collapsed Cluster Plot<a class="anchor" aria-label="anchor" href="#cnet-collapsed-cluster-plot"></a>
</h3>
<p>The Cnet Collapsed Cluster Plot is often the basis for manuscript
figures. The typical workflow is demonstrated below.</p>
<ul>
<li>
<code>cnet &lt;- CnetCollapsed(mpf, type="set")</code> retrieves the
Cnet <code>igraph</code> object.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the cnet</span></span>
<span><span class="va">cnet</span> <span class="op">&lt;-</span> <span class="fu">CnetCollapsed</span><span class="op">(</span><span class="va">Mpf</span>, do_plot<span class="op">=</span><span class="cn">FALSE</span>, type<span class="op">=</span><span class="st">"set"</span><span class="op">)</span>;</span></code></pre></div>
<p><code><a href="../reference/jam_igraph.html">jam_igraph()</a></code> is a custom plotting function with
enhancements:</p>
<ul>
<li>
<code>node_factor=2</code> multiplies node size by 2.</li>
<li>
<code>label_dist_factor=2</code> multiplies label distance from node
center by 5.</li>
<li>
<code>use_shadowText=TRUE</code> uses shadowing around the text
labels.</li>
<li>
<code>label_factor_l</code> resizes the node labels by ‘nodeType’
for Gene and Set.</li>
<li>It applies edge bundling, which helps with large networks.</li>
<li>It plots using vectorized optimization.</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># jam_graph instead of plot()</span></span>
<span><span class="fu"><a href="../reference/jam_igraph.html">jam_igraph</a></span><span class="op">(</span><span class="va">cnet</span>,</span>
<span>   node_factor<span class="op">=</span><span class="fl">2</span>,</span>
<span>   use_shadowText<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>   label_dist_factor<span class="op">=</span><span class="fl">5</span>,</span>
<span>   label_factor_l<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nodeType<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Gene<span class="op">=</span><span class="fl">2</span>, Set<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="importClusterProfiler_files/figure-html/cnet-collapsed-1.png" alt="Cnet collapsed plot suitable for customization." width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="shinycat-for-custom-cnet-layout">ShinyCat for Custom Cnet Layout<a class="anchor" aria-label="anchor" href="#shinycat-for-custom-cnet-layout"></a>
</h2>
<p>The R-shiny Cnet Adjustment Tool <strong>ShinyCat</strong> is
intended to help polish the Cnet plot layout when making a final
figure.</p>
<p>The R-shiny app uses several functions:</p>
<ul>
<li>
<code><a href="../reference/nudge_igraph_node.html">nudge_igraph_node()</a></code>: mode individual nodes</li>
<li>
<code><a href="../reference/adjust_cnet_nodeset.html">adjust_cnet_nodeset()</a></code>: adjust spacing, position,
rotation of a nodeset</li>
<li>
<code><a href="../reference/reorderIgraphNodes.html">reorder_igraph_nodes()</a></code>: sort nodes in a group by
color</li>
<li>
<code><a href="../reference/spread_igraph_labels.html">spread_igraph_labels()</a></code>: arrage labels radially away
from incoming edges</li>
<li>
<code><a href="../reference/bulk_cnet_adjustments.html">bulk_cnet_adjustments()</a></code>: several operations applied in
bulk</li>
</ul>
<blockquote>
<p>Make sure to assign the output to a variable, or to click “Save
RData” from within the R-shiny app. For example:</p>
</blockquote>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">output_env</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/launch_shinycat.html">launch_shinycat</a></span><span class="op">(</span>g<span class="op">=</span><span class="va">cnet</span><span class="op">)</span></span></code></pre></div>
<p>The output is stored in an <code>environment</code> called
<code>output_env</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># obtain the output data</span></span>
<span><span class="va">adj_cnet</span> <span class="op">&lt;-</span> <span class="va">output_env</span><span class="op">$</span><span class="va">adj_cnet</span>;</span></code></pre></div>
<p>Then the new Cnet plot can be plotted, for example:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># jam_graph</span></span>
<span><span class="fu"><a href="../reference/jam_igraph.html">jam_igraph</a></span><span class="op">(</span><span class="va">adj_cnet</span>,</span>
<span>   node_factor<span class="op">=</span><span class="fl">2</span>,</span>
<span>   use_shadowText<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>   label_factor_l<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>nodeType<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>Gene<span class="op">=</span><span class="fl">2</span>, Set<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="shinycat-screenshot">ShinyCat Screenshot<a class="anchor" aria-label="anchor" href="#shinycat-screenshot"></a>
</h3>
<p>An example of ShinyCat in action is shown below.</p>
<p><img src="shinycat-2.png" alt="Screenshot of ShinyCat in action, with a Cnet network plot in the center, and several inputs on the left to adjust the layout." width="100%" style="display: block; margin: auto;"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
